{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ca81bc7f",
   "metadata": {},
   "source": [
    "# ARCHS4 model building with CLAMP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1b972c8",
   "metadata": {},
   "source": [
    "ðŸ’¡ **Environment:** `clamp-analyses`  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48ba7da3",
   "metadata": {},
   "source": [
    "This notebook focuses on preprocessing data from the ARCHS4 database. It covers initial cleaning, and preparation for subsequent analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be805cf1",
   "metadata": {},
   "source": [
    "## Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4ed5e6f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(hdf5r)\n",
    "library(biomaRt)\n",
    "library(dplyr)\n",
    "library(here)\n",
    "library(CLAMP)\n",
    "\n",
    "source(here(\"config.R\"))\n",
    "set.seed(config$ARCHS4$RANDOM_SVD_SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc858c79",
   "metadata": {},
   "source": [
    "## Output directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "02f368b4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "output_dir <- config$ARCHS4$DATASET_FOLDER\n",
    "dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7680eee5",
   "metadata": {},
   "source": [
    "## Preprocess archs4 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "98911899",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "file_path <- config$ARCHS4$DATASET_FILE\n",
    "\n",
    "h5        <- H5File$new(file_path, mode = \"r\")\n",
    "dset      <- h5[[\"/data/expression\"]]\n",
    "gene_symbols <- h5[[\"/meta/genes/symbol\"]]$read()\n",
    "gene_ids     <- h5[[\"/meta/genes/ensembl_gene\"]]$read()\n",
    "sample_names <- h5[[\"/meta/samples/geo_accession\"]]$read()\n",
    "sc_samples      <- h5[[\"/meta/samples/singlecellprobability\"]]$read()\n",
    "lib_strategy <- h5[[\"/meta/samples/library_strategy\"]]$read()     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6c49b65a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "samples_idx <- which(sc_samples < 0.5)\n",
    "sample_names_n_singlecell  <- sample_names[samples_idx]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "260c3378",
   "metadata": {},
   "source": [
    "Get genes length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "328654d0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'hgnc_symbol'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "output_file <- file.path(output_dir, \"gene_lengths.rds\")\n",
    "\n",
    "ensembl_ver <- config$ARCHS4$DATASET_ENSEMBL_VERSION\n",
    "ensembl <- useEnsembl(biomart=\"ensembl\", version=ensembl_ver)\n",
    "mart <- biomaRt::useDataset(\"hsapiens_gene_ensembl\", ensembl)\n",
    "\n",
    "gene_info <- biomaRt::getBM(\n",
    "  filters = \"hgnc_symbol\",\n",
    "  attributes = c(\"ensembl_gene_id\", \"hgnc_symbol\", \"transcript_length\"),\n",
    "  values = gene_symbols, \n",
    "  mart = mart\n",
    ")\n",
    "\n",
    "gene_lengths <- gene_info %>%\n",
    "  group_by(hgnc_symbol, ensembl_gene_id) %>%\n",
    "  summarize(gene_length_bp = max(transcript_length, na.rm = TRUE)) %>%\n",
    "  pull(gene_length_bp, name = hgnc_symbol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5e7d7257",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# keep in data only genes for which we have lengths\n",
    "# here gene symbols are duplicated (we'll use that later)\n",
    "gene_symbols_idx <- which(gene_symbols %in% names(gene_lengths) )\n",
    "gene_symbols <- gene_symbols[gene_symbols_idx]\n",
    "\n",
    "# get unique gene symbols\n",
    "# since we are gonna use rowsum to aggregate same gene symbols, I use it here\n",
    "# to get a list of unique and fixed gene symbols\n",
    "summed <- rowsum(t(dset[1:10, gene_symbols_idx]), group = gene_symbols)\n",
    "gene_symbols_unique <- rownames(summed)\n",
    "stopifnot(length(gene_symbols_unique) == length(unique(gene_symbols)))\n",
    "gene_lengths <- gene_lengths[gene_symbols_unique]\n",
    "n_genes <- length(gene_symbols_unique)\n",
    "rm(summed)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff4bc66e",
   "metadata": {},
   "source": [
    "## FMB matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "7c4fe23e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "âœ… Removed existing FBM file: /home/msubirana/Documents/pivlab/clamp-analyses/output/archs4/FBMarchs4.bk\n",
      "\n",
      "â„¹ï¸ Not found, skipping: /home/msubirana/Documents/pivlab/clamp-analyses/output/archs4/FBMarchs4_preproc.bk\n",
      "\n",
      "â„¹ï¸ Not found, skipping: /home/msubirana/Documents/pivlab/clamp-analyses/output/archs4/FBMarchs4_filtered.bk\n",
      "\n",
      "â„¹ï¸ Not found, skipping: /home/msubirana/Documents/pivlab/clamp-analyses/output/archs4/FBMarchs4_preproc_filtered.bk\n",
      "\n",
      "â„¹ï¸ Not found, skipping: /home/msubirana/Documents/pivlab/clamp-analyses/output/archs4/FBMarchs4_preproc_filtered.rds\n",
      "\n",
      "âœ… Removed existing FBM file: /home/msubirana/Documents/pivlab/clamp-analyses/output/archs4/FBMarchs4_nosinglecell.bk\n",
      "\n",
      "âœ… Removed existing FBM file: /home/msubirana/Documents/pivlab/clamp-analyses/output/archs4/FBMarchs4_nosinglecell_preproc.bk\n",
      "\n",
      "âœ… Removed existing FBM file: /home/msubirana/Documents/pivlab/clamp-analyses/output/archs4/FBMarchs4_nosinglecell_preproc_filtered.bk\n",
      "\n",
      "â„¹ï¸ Not found, skipping: /home/msubirana/Documents/pivlab/clamp-analyses/output/archs4/FBMarchs4_nosinglecell_preproc_filtered.rds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "fbm_files <- c(\n",
    "  file.path(output_dir, \"FBMarchs4.bk\"),\n",
    "  file.path(output_dir, \"FBMarchs4_preproc.bk\"),\n",
    "  file.path(output_dir, \"FBMarchs4_filtered.bk\"),\n",
    "  file.path(output_dir, \"FBMarchs4_preproc_filtered.bk\"),\n",
    "  file.path(output_dir, \"FBMarchs4_preproc_filtered.rds\"),\n",
    "  file.path(output_dir, \"FBMarchs4_nosinglecell.bk\"),\n",
    "  file.path(output_dir, \"FBMarchs4_nosinglecell_preproc.bk\"),\n",
    "  file.path(output_dir, \"FBMarchs4_nosinglecell_preproc_filtered.bk\"),\n",
    "  file.path(output_dir, \"FBMarchs4_nosinglecell_preproc_filtered.rds\")\n",
    ")\n",
    "\n",
    "for (f in fbm_files) {\n",
    "  if (file.exists(f)) {\n",
    "    res <- unlink(f)\n",
    "    if (identical(res, 0L)) {\n",
    "      message(\"âœ… Removed existing FBM file: \", f)\n",
    "    } else {\n",
    "      warning(\"âš ï¸ Failed to remove: \", f)\n",
    "    }\n",
    "  } else {\n",
    "    message(\"â„¹ï¸ Not found, skipping: \", f)\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3b51798",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#remove just for testing\n",
    "n_genes <- length(gene_symbols_unique)\n",
    "n_samples <- length(sample_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "0637fd63",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Create the FBM\n",
    "fbm_file <- file.path(output_dir, \"FBMarchs4\")\n",
    "bkfile   <- paste0(fbm_file, \".bk\")\n",
    "\n",
    "archs4FBM <- FBM(\n",
    "    nrow        = n_genes,\n",
    "    ncol        = n_samples,\n",
    "    backingfile = fbm_file,\n",
    "    create_bk   = TRUE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "580460e7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%"
     ]
    }
   ],
   "source": [
    "block_size <- 100\n",
    "n_blocks   <- ceiling(n_samples / block_size)\n",
    "\n",
    "pb <- txtProgressBar(min = 0, max = n_blocks, style = 3)\n",
    "\n",
    "for (i in 1:n_blocks) {\n",
    "  setTxtProgressBar(pb, i)\n",
    "\n",
    "  start_row <- (i-1) * block_size + 1\n",
    "  end_row <- min(i * block_size, n_samples)\n",
    "\n",
    "  raw_block <- NULL\n",
    "  tryCatch(\n",
    "    expr = {\n",
    "      raw_block <- dset[start_row:end_row, ]\n",
    "    },\n",
    "    error = function(e) {\n",
    "      message(paste(\"Error with block: \", i))\n",
    "    }\n",
    "  )\n",
    "  if (is.null(raw_block)) {\n",
    "    # FIXME: I add a dummy block here with near-to-zero values; not sure if\n",
    "    # this is the best approach here; for some reason, there is a block that\n",
    "    # cannot be read in ARCHS4, so I \"skip\" it here\n",
    "    raw_block <- matrix(1e-10, (end_row - start_row + 1), n_genes)\n",
    "  }\n",
    "\n",
    "  # subset genes to those with gene length only\n",
    "  raw_block <- raw_block[, gene_symbols_idx]\n",
    "\n",
    "  # aggregate duplicated gene symbols\n",
    "  raw_block_t <- t(raw_block)\n",
    "  raw_block_summed <- rowsum(raw_block_t, group = gene_symbols)\n",
    "  raw_block <- t(raw_block_summed)\n",
    "  rownames(raw_block) <- NULL\n",
    "\n",
    "  # tpm normalize\n",
    "\n",
    "  tpm_norm <- function(counts, gene.lengths) {\n",
    "    if (!is.matrix(counts)) {\n",
    "      stop(\"`counts` must be a matrix.\")\n",
    "    }\n",
    "    counts <- as.matrix(counts)\n",
    "    \n",
    "    if (is.null(names(gene.lengths))) {\n",
    "      stop(\"`gene.lengths` must be a named numeric vector.\")\n",
    "    }\n",
    "    if (!all(colnames(counts) %in% names(gene.lengths))) {\n",
    "      stop(\"All column names of `counts` must be in names(gene.lengths).\")\n",
    "    }\n",
    "    \n",
    "    # Reorder lengths to match columns\n",
    "    #lengths.bp <- gene.lengths[colnames(counts)]\n",
    "    \n",
    "    # Convert lengths to kilobases\n",
    "    lengths.kb <- gene.lengths / 1e3\n",
    "    \n",
    "    # 1) Divide counts by gene length in kilobases â†’ reads per kilobase\n",
    "    rpk <- sweep(counts, 2, lengths.kb, FUN = \"/\")\n",
    "    \n",
    "    # 2) Compute per-sample scaling factor: sum of RPKs\n",
    "    per.sample.sum <- rowSums(rpk)\n",
    "    \n",
    "    # 3) Divide RPKs by the sum and multiply by 1e6 â†’ TPM\n",
    "    tpm <- sweep(rpk, 1, per.sample.sum, FUN = \"/\") * 1e6\n",
    "    \n",
    "    return(tpm)\n",
    "  }\n",
    "\n",
    "  raw_block <- t(tpm_norm(raw_block, gene_lengths))\n",
    "  archs4FBM[, start_row:end_row] <- as.matrix(raw_block)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9934d5ec",
   "metadata": {},
   "source": [
    "Once done, close h5 file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a470a0d0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "h5$close_all()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62f6b22c",
   "metadata": {},
   "source": [
    "## ARCHS4 preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "480b5893",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# remove single cell samples\n",
    "samples_idx <- which(sample_names %in% sample_names_n_singlecell)\n",
    "sample_names <- sample_names[samples_idx]\n",
    "n_samples <- length(sample_names)\n",
    "\n",
    "archs4FBM <- big_copy(\n",
    "  archs4FBM,\n",
    "  ind.col = samples_idx,\n",
    "  backingfile = file.path(output_dir, \"FBMarchs4_nosinglecell\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "676fda6d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Applying log2 transformation\n",
      "\n",
      "Filling NAs with 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Preprocess\n",
    "prep_archs4 <- preprocessCLAMPFBM(\n",
    "  fbm = archs4FBM,\n",
    "  mean_cutoff = config$ARCHS4$GENES_MEAN_CUTOFF,\n",
    "  var_cutoff = config$ARCHS4$GENES_VAR_CUTOFF,\n",
    "  block_size = block_size\n",
    ")\n",
    "\n",
    "archs4_fbm_filt <- prep_archs4$fbm_filtered\n",
    "archs4_rowStats <- prep_archs4$rowStats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "f1709dba",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "archs4_genes <- gene_symbols_unique[prep_archs4$kept_rows]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "e09fd62a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Applying Z-score transformation\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Zscore\n",
    "zscoreCLAMPFBM(\n",
    "    fbm_filtered=archs4_fbm_filt,\n",
    "    rowStats=archs4_rowStats,\n",
    "    chunk_size=block_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "c04d93ba",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "saveRDS(prep_archs4, file = file.path(output_dir, \"prep_archs4.rds\"))\n",
    "saveRDS(archs4_genes, file = file.path(output_dir, \"archs4_genes.rds\"))\n",
    "saveRDS(sample_names, file = file.path(output_dir, \"archs4_samples.rds\"))\n",
    "saveRDS(archs4_fbm_filt, file = file.path(output_dir, \"archs4_fbm_filt.rds\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
