{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ca81bc7f",
   "metadata": {},
   "source": [
    "# recount2 model building with CLAMP and PLIER"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3ee54fb",
   "metadata": {},
   "source": [
    "ðŸ’¡ **Environment:** `clamp-analyses`  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56c6fd3c",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e69fa323",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "# CLAMP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be805cf1",
   "metadata": {},
   "source": [
    "## Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec5663f0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Create a timestamp to track the start of the analysis\n",
    "start_time <- Sys.time()\n",
    "cat(\"recount2 CLAMP and PLIER analysis started at:\", format(start_time), \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4ed5e6f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "if (!requireNamespace(\"PLIER\", quietly = TRUE)) {\n",
    "    devtools::install_github(\"wgmao/PLIER\")\n",
    "}\n",
    "\n",
    "library(bigstatsr)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "library(rsvd)\n",
    "library(glmnet)\n",
    "library(Matrix)\n",
    "library(knitr)\n",
    "library(here)\n",
    "library(PLIER)\n",
    "library(CLAMP)\n",
    "\n",
    "source(here(\"config.R\"))\n",
    "\n",
    "set.seed(config$recount2$RANDOM_SVD_SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc858c79",
   "metadata": {},
   "source": [
    "## Output directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "02f368b4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "output_data_dir <- config$recount2$DATASET_FOLDER\n",
    "dir.create(output_data_dir, showWarnings = FALSE, recursive = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7286f749",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "block_size <- config$GENERAL$CHUNK_SIZE\n",
    "N_CORES    <- config$recount2$N_CORES"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf137533",
   "metadata": {},
   "source": [
    "## Download recount2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "759882bd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Required files already exist. Skipping download and decompression.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "url      <- config$recount2$URL\n",
    "out_dir  <- here::here(\"data\", \"recount2\")\n",
    "zip_file <- file.path(out_dir, \"recount2.zip\")\n",
    "rds_file <- file.path(out_dir, \"recount2_PLIER_data\", \"recount_data_prep_PLIER.RDS\")\n",
    "rpkm_file <- file.path(out_dir, \"recount2_PLIER_data\", \"recount_rpkm.RDS\")\n",
    "\n",
    "if (!file.exists(rds_file) && !file.exists(rpkm_file)) {\n",
    "    if (!dir.exists(out_dir)) dir.create(out_dir, recursive = TRUE)\n",
    "    \n",
    "    system(paste('wget', url, '-O', zip_file))\n",
    "    unzip(zip_file, exdir = out_dir)\n",
    "} else {\n",
    "    message(\"Required files already exist. Skipping download and decompression.\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7680eee5",
   "metadata": {},
   "source": [
    "## Preprocess recount2 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "41d0a2d0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Read preprocessed recount2 data\n",
    "data_mat <- readRDS(rds_file)\n",
    "\n",
    "meta <- list()\n",
    "meta$gene_symbols <- rownames(data_mat$rpkm.cm)\n",
    "meta$samples <- colnames(data_mat$rpkm.cm)\n",
    "\n",
    "rm(data_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "8c3dc65e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dplyr::inner_join(gene.df, rpkm.df, by = \"ensembl_gene_id\"):\n",
      "â€œ\u001b[1m\u001b[22mDetected an unexpected many-to-many relationship between `x` and `y`.\n",
      "\u001b[36mâ„¹\u001b[39m Row 23 of `x` matches multiple rows in `y`.\n",
      "\u001b[36mâ„¹\u001b[39m Row 29567 of `y` matches multiple rows in `x`.\n",
      "\u001b[36mâ„¹\u001b[39m If a many-to-many relationship is expected, set `relationship =\n",
      "  \"many-to-many\"` to silence this warning.â€\n"
     ]
    }
   ],
   "source": [
    "# Read in recount2 file\n",
    "\n",
    "rpkm.df <- readRDS(rpkm_file)\n",
    "\n",
    "mart <- biomaRt::useDataset(\"hsapiens_gene_ensembl\",\n",
    "                            biomaRt::useMart(\"ensembl\"))\n",
    "\n",
    "genes <- unlist(lapply(strsplit(rpkm.df$ENSG, \"[.]\"), `[[`, 1))\n",
    "\n",
    "rpkm.df$ensembl_gene_id <- unlist(lapply(strsplit(rpkm.df$ENSG, \"[.]\"),\n",
    "                                         `[[`, 1))\n",
    "                                         \n",
    "gene.df <- biomaRt::getBM(filters = \"ensembl_gene_id\",\n",
    "                          attributes = c(\"ensembl_gene_id\", \"hgnc_symbol\"),\n",
    "                          values = genes,\n",
    "                          mart = mart)\n",
    "\n",
    "gene.df <- gene.df %>% dplyr::filter(complete.cases(.))\n",
    "\n",
    "rpkm.df <- dplyr::inner_join(gene.df, rpkm.df,\n",
    "                             by = \"ensembl_gene_id\")\n",
    "\n",
    "rownames(rpkm.df) <- make.names(rpkm.df$hgnc_symbol, unique = TRUE)\n",
    "\n",
    "rpkm.df <- rpkm.df %>% dplyr::select(-c(ensembl_gene_id:ENSG))\n",
    "\n",
    "data_mat <- rpkm.df[meta$gene_symbols, meta$samples]\n",
    "\n",
    "rm(rpkm.df)\n",
    "\n",
    "data_mat <- as.matrix(as.data.table(data_mat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68af70e3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "fbm_files <- c(\n",
    "  file.path(output_data_dir, \"recount2.bk\"),\n",
    "  file.path(output_data_dir, \"recount2.bk_preproc.bk\"),\n",
    "  file.path(output_data_dir, \"recount2.bk_preproc_filtered.bk\"),\n",
    "  file.path(output_data_dir, \"recount2.bk_preproc_filtered.rds\")\n",
    ")\n",
    "\n",
    "for (f in fbm_files) {\n",
    "  if (file.exists(f)) {\n",
    "    res <- unlink(f)\n",
    "    if (identical(res, 0L)) {\n",
    "      message(\"âœ… Removed existing FBM file: \", f)\n",
    "    } else {\n",
    "      warning(\"âš ï¸ Failed to remove: \", f)\n",
    "    }\n",
    "  } else {\n",
    "    message(\"â„¹ï¸ Not found, skipping: \", f)\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "b2fb30b5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Create the FBM\n",
    "n_genes <- length(meta$gene_symbols)\n",
    "n_samples <- length(meta$samples)\n",
    "\n",
    "fbm_file <- file.path(output_data_dir, \"FBMrecount2\")\n",
    "recount2FBM <- FBM(nrow = n_genes, ncol = n_samples, backingfile = fbm_file, create_bk = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62a459dd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Populate it with data\n",
    "block_size <- config$GENERAL$CHUNK_SIZE\n",
    "n_blocks <- ceiling(n_genes / block_size)\n",
    "\n",
    "for (i in 1:n_blocks) {\n",
    "  start_row <- (i-1) * block_size + 1\n",
    "  end_row <- min(i * block_size, nrow(data_mat))\n",
    "  recount2FBM[start_row:end_row, ] <- as.matrix(data_mat[start_row:end_row, ])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfe9908d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Applying log2 transformation\n",
      "\n",
      "Filling NAs with 0\n",
      "\n",
      "Applying Z-score transformation\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Preprocess and zâ€‘score FBM\n",
    "prep_recount2 <- preprocessCLAMPFBM(\n",
    "  fbm        = recount2FBM,\n",
    "  mean_cutoff= config$recount2$GENES_MEAN_CUTOFF,\n",
    "  var_cutoff = config$recount2$GENES_VAR_CUTOFF\n",
    ")\n",
    "\n",
    "recount2_fbm_filt <- prep_recount2$fbm_filtered\n",
    "recount2_rowStats <- prep_recount2$rowStats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c2ec872",
   "metadata": {},
   "source": [
    "Check overlap genes with recount2 RNA-seq data and TWAS gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14dea818",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "twas_genes <- read.csv(here('output/creating_twas_gwas_list/gene_list_union_mashr_and_elastic_net_with_phi.csv'))\n",
    "overlap_genes <- intersect(genes[prep_recount2$kept_rows], twas_genes$Gene_Symbol)\n",
    "\n",
    "percent_overlap <- (length(overlap_genes) / length(twas_genes$Gene_Symbol)) * 100\n",
    "\n",
    "cat(\"Number of overlapping genes:\", length(overlap_genes), \"\\n\")\n",
    "cat(\"Percentage overlap (relative to TWAS genes):\", round(percent_overlap, 2), \"%\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee0ee831",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "zscoreCLAMPFBM(recount2_fbm_filt, recount2_rowStats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "zscoreCLAMPFBM(recount2_fbm_filt, recount2_rowStats, ncores=N_CORES)\n",
    "recount2_genes <- genes[prep_recount2$kept_rows]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "saveRDS(samples, file = file.path(output_data_dir, \"recount2_samples.rds\"))\n",
    "saveRDS(recount2_genes, file = file.path(output_data_dir, \"recount2_genes.rds\"))\n",
    "saveRDS(recount2_fbm_filt, file = file.path(output_data_dir, \"recount2_fbm_filt.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "df_recount2_fbm_filt <- as.data.frame(as.matrix(recount2_fbm_filt[]))\n",
    "colnames(df_recount2_fbm_filt) <- samples\n",
    "rownames(df_recount2_fbm_filt) <- recount2_genes\n",
    "head(df_recount2_fbm_filt)\n",
    "\n",
    "saveRDS(df_recount2_fbm_filt, file = file.path(output_data_dir, \"df_recount2_fbm_filt.rds\"))\n",
    "write.csv(df_recount2_fbm_filt, file = file.path(output_data_dir, \"df_recount2_fbm_filt.csv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62186aa8",
   "metadata": {},
   "source": [
    "## SVD computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e3990c6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "if (!file.exists(file.path(output_data_dir, \"recount2_svdRes.rds\"))) {\n",
    "\n",
    "  g_fb <- nrow(recount2_fbm_filt)\n",
    "  samples_fb <- ncol(recount2_fbm_filt)\n",
    "  SVD_K_recount2 <- min(g_fb, samples_fb) - 1\n",
    "\n",
    "  message(\"Using SVD K = \", SVD_K_recount2)\n",
    "\n",
    "  if (N_CORES > 1) {\n",
    "    # disable BLAS parallelization when parallelizing\n",
    "    options(bigstatsr.check.parallel.blas = FALSE)\n",
    "    blas_nproc <- getOption(\"default.nproc.blas\")\n",
    "    options(default.nproc.blas = NULL)\n",
    "  }\n",
    "\n",
    "  recount2_svdRes <- big_randomSVD(\n",
    "    recount2_fbm_filt,\n",
    "    k = SVD_K_recount2,\n",
    "    ncores = N_CORES\n",
    "  )\n",
    "\n",
    "  saveRDS(recount2_svdRes, file = file.path(output_data_dir, \"recount2_svdRes.rds\"))\n",
    "\n",
    "  if (N_CORES > 1) {\n",
    "    # restore previous state\n",
    "    options(bigstatsr.check.parallel.blas = TRUE)\n",
    "    options(default.nproc.blas = blas_nproc)\n",
    "  }\n",
    "\n",
    "} else {\n",
    "  message(\"recount2_svdRes already exists, skipping SVD computation.\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dedb1810",
   "metadata": {},
   "source": [
    "## Estimate K for CLAMP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52d08c85",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Inferred PLIER K = 129\n",
      "\n"
     ]
    }
   ],
   "source": [
    "recount2_fbm_filt <- readRDS(file.path(output_data_dir, \"recount2_fbm_filt.rds\"))\n",
    "recount2_svdRes <- readRDS(file.path(output_data_dir, \"recount2_svdRes.rds\"))\n",
    "\n",
    "CLAMP_K_recount2 <- num.pc(list(d = recount2_svdRes$d)) * 2\n",
    "message(\"Inferred CLAMP K = \", CLAMP_K_recount2)\n",
    "\n",
    "saveRDS(CLAMP_K_recount2, file = file.path(output_data_dir, \"CLAMP_K_recount2.rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cdcb0f",
   "metadata": {},
   "source": [
    "## CLAMPbase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dc460fb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "****\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"L1 is set to 96.4556873352599\"\n",
      "[1] \"L2 is set to 289.36706200578\"\n",
      "Progress 48 / 200 | Bdiff=0.000489, minCor=0.997171Converged at iteration= 48 | Bdiff=0.000489,  tol=0.000500     "
     ]
    }
   ],
   "source": [
    "recount2_baseRes <- CLAMPbase(\n",
    "  Y      = recount2_fbm_filt,\n",
    "  svdres = recount2_svdRes,\n",
    "  trace  = TRUE,\n",
    "  clamp_k = CLAMP_K_recount2\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "recount2_baseRes$Z <- data.frame(recount2_baseRes$Z)\n",
    "rownames(recount2_baseRes$Z) <- recount2_genes\n",
    "head(recount2_baseRes$Z)\n",
    "\n",
    "recount2_baseRes$B <- data.frame(recount2_baseRes$B)\n",
    "colnames(recount2_baseRes$B) <- samples\n",
    "head(recount2_baseRes$B)\n",
    "\n",
    "saveRDS(recount2_baseRes, file = file.path(output_data_dir, \"CLAMPbase.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "model_dir <- file.path(output_data_dir, \"CLAMPbase\")\n",
    "dir.create(model_dir, showWarnings = FALSE, recursive = TRUE)\n",
    "\n",
    "B <- recount2_baseRes$B\n",
    "write.csv(B, file.path(model_dir, \"B.csv\"))\n",
    "\n",
    "Z <- recount2_baseRes$Z\n",
    "write.csv(Z, file.path(model_dir, \"Z.csv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ee267bb",
   "metadata": {},
   "source": [
    "## Prepare pathway priors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "960031ac",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "recount2_gmtList <- list(\n",
    "  KEGG = getGMT(\"https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName=KEGG_2021_Human\"),\n",
    "  BP = getGMT(\"https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName=GO_Biological_Process_2025\"),\n",
    "  GTEx_Tissues = getGMT(\"https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName=GTEx_Tissues_V8_2023\")\n",
    ")\n",
    "\n",
    "# prefix each geneâ€set name with its library to guarantee uniqueness\n",
    "for(lib in names(recount2_gmtList)) {\n",
    "  names(recount2_gmtList[[lib]]) <- paste0(lib, \"_\", names(recount2_gmtList[[lib]]))\n",
    "}\n",
    "\n",
    "recount2_pathMat <- gmtListToSparseMat(recount2_gmtList)\n",
    "recount2_matched <- getMatchedPathwayMat(recount2_pathMat, recount2_genes)\n",
    "recount2_chatObj <- getChat(recount2_matched)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6526304d",
   "metadata": {},
   "source": [
    "## CLAMPfull"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5673ed3a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**PLIER v2 **\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "using provided PLIERbase result\n",
      "\n",
      "L1=96.4556873352599; L2=289.36706200578\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converged at 108 / 350 | Bdiff=0.000011, minCor=0.999752olumns is 98, Number of annotated columns is 98, Number of annotated columns is 99, Number of annotated columns is 100, Number of annotated columns is 100, Number of annotated columns is 94, Number of annotated columns is 95, Number of annotated columns is 95, Number of annotated columns is 95, Number of annotated columns is 97, Number of annotated columns is 93, Number of annotated columns is 94, Number of annotated columns is 95, Number of annotated columns is 95, Number of annotated columns is 95, Number of annotated columns is 94, Number of annotated columns is 95, Number of annotated columns is 96, Number of annotated columns is 97, Number of annotated columns is 97, Number of annotated columns is 95, Number of annotated columns is 95, Number of annotated columns is 95, Number of annotated columns is 95, Number of annotated columns is 96, Number of annotated columns is 93, Number of annotated columns is 93, Number of annotated columns is 94, Number of annotated columns is 94, Number of annotated columns is 94, Number of annotated columns is 91, Number of annotated columns is 94, Number of annotated columns is 94, Number of annotated columns is 94, Number of annotated columns is 94, Number of annotated columns is 93, Number of annotated columns is 93, Number of annotated columns is 94, Number of annotated columns is 94, Number of annotated columns is 94, Number of annotated columns is 92, Number of annotated columns is 92, Number of annotated columns is 93, Number of annotated columns is 93, Number of annotated columns is 93, Number of annotated columns is 93, Number of annotated columns is 93, Number of annotated columns is 93, Number of annotated columns is 93, Number of annotated columns is 93\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating Z for CV\n",
      "\n",
      "crossValidation\n",
      "\n",
      "There are 68  LVs with AUC>0.70\n",
      "\n",
      "There are 32  LVs with AUC>0.90\n",
      "\n"
     ]
    }
   ],
   "source": [
    "recount2_fullRes <- CLAMPfull(\n",
    "    Y = recount2_fbm_filt,\n",
    "    svdres = recount2_svdRes,\n",
    "    priorMat = recount2_matched,\n",
    "    clamp.base.result = recount2_baseRes,\n",
    "    use_cpp = TRUE,\n",
    "    trace = TRUE,\n",
    "    clamp_k = CLAMP_K_recount2\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2a0236d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "recount2_baseRes$Z <- data.frame(recount2_baseRes$Z)\n",
    "rownames(recount2_baseRes$Z) <- recount2_genes\n",
    "head(recount2_baseRes$Z)\n",
    "\n",
    "recount2_baseRes$B <- data.frame(recount2_baseRes$B)\n",
    "colnames(recount2_baseRes$B) <- samples\n",
    "head(recount2_baseRes$B)\n",
    "\n",
    "recount2_fullRes$summary <- recount2_fullRes$summary %>%\n",
    "    dplyr::rename(LV = LV_index)  %>% \n",
    "    dplyr::mutate(LV = paste0('LV', LV))\n",
    "\n",
    "saveRDS(recount2_fullRes, file = file.path(output_data_dir, \"CLAMPfull.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "model_dir <- file.path(output_data_dir, \"CLAMPfull\")\n",
    "dir.create(model_dir, showWarnings = FALSE, recursive = TRUE)\n",
    "\n",
    "B <- recount2_fullRes$B\n",
    "write.csv(B, file.path(model_dir, \"B.csv\"))\n",
    "\n",
    "Z <- recount2_fullRes$Z\n",
    "write.csv(Z, file.path(model_dir, \"Z.csv\"))\n",
    "\n",
    "summary <- recount2_fullRes$summary\n",
    "write.csv(summary, file.path(model_dir, \"summary.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Create a timestamp to track the end of the analysis\n",
    "end_time <- Sys.time()\n",
    "elapsed_time <- end_time - start_time\n",
    "cat(\"Recount2 CLAMP analysis completed at:\", format(end_time), \"\\n\")\n",
    "cat(\"Total elapsed time:\", format(elapsed_time), \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "947ca517",
   "metadata": {},
   "source": [
    "# PLIER"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb463b2b",
   "metadata": {},
   "source": [
    "Run PLIER with the same inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95a907e4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Removing 0 pathways with too few genes\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 289.3671\n",
      "[1] \"L2 is set to 289.36706200578\"\n",
      "[1] \"L1 is set to 144.68353100289\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "errorY (SVD based:best possible) = 0.8092\n",
      "\n",
      "New L3 is 0.00218749111818289\n",
      "\n",
      "New L3 is 0.00193045413622771\n",
      "\n",
      "New L3 is 0.00170361979580257\n",
      "\n",
      "New L3 is 0.00193045413622771\n",
      "\n",
      "New L3 is 0.00218749111818289\n",
      "\n",
      "New L3 is 0.00218749111818289\n",
      "\n",
      "New L3 is 0.00218749111818289\n",
      "\n",
      "New L3 is 0.00218749111818289\n",
      "\n",
      "New L3 is 0.00218749111818289\n",
      "\n",
      "New L3 is 0.00193045413622771\n",
      "\n",
      "New L3 is 0.00218749111818289\n",
      "\n",
      "Bdiff is not decreasing\n",
      "\n",
      "Bdiff is not decreasing\n",
      "\n",
      "Bdiff is not decreasing\n",
      "\n",
      "Bdiff is not decreasing\n",
      "\n",
      "Bdiff is not decreasing\n",
      "\n",
      "Bdiff is not decreasing\n",
      "\n",
      "converged at  iteration 234 Bdiff is not decreasing\n",
      "\n",
      "There are 53  LVs with AUC>0.70\n",
      "\n"
     ]
    }
   ],
   "source": [
    "recount2_plier = PLIER::PLIER(\n",
    "    recount2_fbm_filt[], \n",
    "    as.matrix(recount2_matched), \n",
    "    svdres = recount2_svdRes, \n",
    "    Chat = as.matrix(recount2_chatObj), \n",
    "    doCrossval = TRUE, \n",
    "    k = CLAMP_K_recount2\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 Ã— 37032 of type dbl</caption>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>LV 1</th><td> 3.038562e-11</td><td>-1.164697e-13</td><td> 1.499369e-11</td><td> 2.420048e-11</td><td> 1.076099e-11</td><td> 2.400729e-11</td><td> 2.450469e-11</td><td> 2.529211e-11</td><td> 2.492415e-11</td><td> 2.411303e-11</td><td>â‹¯</td><td>-2.252487e-14</td><td>-9.868426e-13</td><td> 7.459374e-13</td><td>-1.238694e-13</td><td>-5.471264e-13</td><td> 3.208637e-13</td><td>-1.203242e-12</td><td>-9.133111e-13</td><td>-7.087071e-13</td><td>-7.032934e-13</td></tr>\n",
       "\t<tr><th scope=row>2,Cells - Cultured Fibroblasts Female 70-79 Up</th><td>-3.281172e-02</td><td>-3.819620e-02</td><td>-4.021169e-02</td><td>-4.102633e-02</td><td>-3.925897e-02</td><td>-3.957653e-02</td><td>-9.111154e-03</td><td>-7.944571e-03</td><td>-1.111526e-02</td><td>-5.022869e-03</td><td>â‹¯</td><td> 3.445032e-02</td><td> 9.405888e-02</td><td> 5.244139e-02</td><td> 9.127825e-02</td><td> 8.897180e-02</td><td> 6.534399e-02</td><td> 7.464888e-02</td><td> 1.037498e-01</td><td> 4.380141e-02</td><td> 9.150614e-02</td></tr>\n",
       "\t<tr><th scope=row>3,mRNA Splicing, via Spliceosome (GO:0000398)</th><td>-9.481710e-02</td><td>-7.740678e-02</td><td>-9.260198e-02</td><td>-8.277387e-02</td><td>-7.081893e-02</td><td>-8.502113e-02</td><td>-6.194781e-02</td><td>-6.563719e-02</td><td>-6.688297e-02</td><td>-7.041583e-02</td><td>â‹¯</td><td>-5.538873e-02</td><td>-6.185108e-02</td><td>-3.681471e-02</td><td>-3.960329e-02</td><td>-6.117308e-02</td><td>-2.970421e-02</td><td>-8.491269e-02</td><td>-3.287660e-02</td><td>-2.841848e-02</td><td>-1.169766e-01</td></tr>\n",
       "\t<tr><th scope=row>4,Translation (GO:0006412)</th><td>-1.167416e-01</td><td>-1.169561e-01</td><td>-1.241410e-01</td><td>-1.144214e-01</td><td>-1.081250e-01</td><td>-1.183666e-01</td><td>-1.348015e-01</td><td>-1.444428e-01</td><td>-1.385527e-01</td><td>-1.426251e-01</td><td>â‹¯</td><td> 5.650564e-02</td><td> 2.639246e-02</td><td> 1.109185e-01</td><td> 6.937041e-02</td><td> 8.791064e-02</td><td> 1.055854e-01</td><td> 1.084018e-01</td><td> 7.893540e-02</td><td> 7.554681e-02</td><td> 1.749875e-01</td></tr>\n",
       "\t<tr><th scope=row>5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)</th><td>-1.475776e-02</td><td>-2.136897e-02</td><td>-2.183746e-02</td><td>-1.868801e-02</td><td>-1.880394e-02</td><td>-1.911732e-02</td><td>-2.733048e-02</td><td>-3.572729e-02</td><td>-3.150585e-02</td><td>-2.717683e-02</td><td>â‹¯</td><td> 4.954903e-03</td><td> 2.586807e-02</td><td> 2.185875e-02</td><td> 4.275850e-02</td><td> 4.369622e-02</td><td> 5.298591e-02</td><td>-1.144091e-02</td><td> 5.217841e-02</td><td>-1.841314e-02</td><td> 2.890525e-02</td></tr>\n",
       "\t<tr><th scope=row>LV 6</th><td> 1.137411e-02</td><td> 1.100171e-02</td><td> 2.021871e-02</td><td> 2.901634e-02</td><td> 1.232140e-02</td><td> 1.205028e-02</td><td> 9.152861e-03</td><td> 1.060361e-02</td><td>-6.413729e-04</td><td> 6.352040e-03</td><td>â‹¯</td><td>-7.302503e-02</td><td>-5.922457e-02</td><td>-7.198972e-02</td><td>-5.402172e-02</td><td>-6.231897e-02</td><td>-6.095302e-02</td><td>-5.806830e-02</td><td>-6.060223e-02</td><td>-7.126057e-02</td><td>-5.891329e-02</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 Ã— 37032 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "\tLV 1 &  3.038562e-11 & -1.164697e-13 &  1.499369e-11 &  2.420048e-11 &  1.076099e-11 &  2.400729e-11 &  2.450469e-11 &  2.529211e-11 &  2.492415e-11 &  2.411303e-11 & â‹¯ & -2.252487e-14 & -9.868426e-13 &  7.459374e-13 & -1.238694e-13 & -5.471264e-13 &  3.208637e-13 & -1.203242e-12 & -9.133111e-13 & -7.087071e-13 & -7.032934e-13\\\\\n",
       "\t2,Cells - Cultured Fibroblasts Female 70-79 Up & -3.281172e-02 & -3.819620e-02 & -4.021169e-02 & -4.102633e-02 & -3.925897e-02 & -3.957653e-02 & -9.111154e-03 & -7.944571e-03 & -1.111526e-02 & -5.022869e-03 & â‹¯ &  3.445032e-02 &  9.405888e-02 &  5.244139e-02 &  9.127825e-02 &  8.897180e-02 &  6.534399e-02 &  7.464888e-02 &  1.037498e-01 &  4.380141e-02 &  9.150614e-02\\\\\n",
       "\t3,mRNA Splicing, via Spliceosome (GO:0000398) & -9.481710e-02 & -7.740678e-02 & -9.260198e-02 & -8.277387e-02 & -7.081893e-02 & -8.502113e-02 & -6.194781e-02 & -6.563719e-02 & -6.688297e-02 & -7.041583e-02 & â‹¯ & -5.538873e-02 & -6.185108e-02 & -3.681471e-02 & -3.960329e-02 & -6.117308e-02 & -2.970421e-02 & -8.491269e-02 & -3.287660e-02 & -2.841848e-02 & -1.169766e-01\\\\\n",
       "\t4,Translation (GO:0006412) & -1.167416e-01 & -1.169561e-01 & -1.241410e-01 & -1.144214e-01 & -1.081250e-01 & -1.183666e-01 & -1.348015e-01 & -1.444428e-01 & -1.385527e-01 & -1.426251e-01 & â‹¯ &  5.650564e-02 &  2.639246e-02 &  1.109185e-01 &  6.937041e-02 &  8.791064e-02 &  1.055854e-01 &  1.084018e-01 &  7.893540e-02 &  7.554681e-02 &  1.749875e-01\\\\\n",
       "\t5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) & -1.475776e-02 & -2.136897e-02 & -2.183746e-02 & -1.868801e-02 & -1.880394e-02 & -1.911732e-02 & -2.733048e-02 & -3.572729e-02 & -3.150585e-02 & -2.717683e-02 & â‹¯ &  4.954903e-03 &  2.586807e-02 &  2.185875e-02 &  4.275850e-02 &  4.369622e-02 &  5.298591e-02 & -1.144091e-02 &  5.217841e-02 & -1.841314e-02 &  2.890525e-02\\\\\n",
       "\tLV 6 &  1.137411e-02 &  1.100171e-02 &  2.021871e-02 &  2.901634e-02 &  1.232140e-02 &  1.205028e-02 &  9.152861e-03 &  1.060361e-02 & -6.413729e-04 &  6.352040e-03 & â‹¯ & -7.302503e-02 & -5.922457e-02 & -7.198972e-02 & -5.402172e-02 & -6.231897e-02 & -6.095302e-02 & -5.806830e-02 & -6.060223e-02 & -7.126057e-02 & -5.891329e-02\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 Ã— 37032 of type dbl\n",
       "\n",
       "| LV 1 |  3.038562e-11 | -1.164697e-13 |  1.499369e-11 |  2.420048e-11 |  1.076099e-11 |  2.400729e-11 |  2.450469e-11 |  2.529211e-11 |  2.492415e-11 |  2.411303e-11 | â‹¯ | -2.252487e-14 | -9.868426e-13 |  7.459374e-13 | -1.238694e-13 | -5.471264e-13 |  3.208637e-13 | -1.203242e-12 | -9.133111e-13 | -7.087071e-13 | -7.032934e-13 |\n",
       "| 2,Cells - Cultured Fibroblasts Female 70-79 Up | -3.281172e-02 | -3.819620e-02 | -4.021169e-02 | -4.102633e-02 | -3.925897e-02 | -3.957653e-02 | -9.111154e-03 | -7.944571e-03 | -1.111526e-02 | -5.022869e-03 | â‹¯ |  3.445032e-02 |  9.405888e-02 |  5.244139e-02 |  9.127825e-02 |  8.897180e-02 |  6.534399e-02 |  7.464888e-02 |  1.037498e-01 |  4.380141e-02 |  9.150614e-02 |\n",
       "| 3,mRNA Splicing, via Spliceosome (GO:0000398) | -9.481710e-02 | -7.740678e-02 | -9.260198e-02 | -8.277387e-02 | -7.081893e-02 | -8.502113e-02 | -6.194781e-02 | -6.563719e-02 | -6.688297e-02 | -7.041583e-02 | â‹¯ | -5.538873e-02 | -6.185108e-02 | -3.681471e-02 | -3.960329e-02 | -6.117308e-02 | -2.970421e-02 | -8.491269e-02 | -3.287660e-02 | -2.841848e-02 | -1.169766e-01 |\n",
       "| 4,Translation (GO:0006412) | -1.167416e-01 | -1.169561e-01 | -1.241410e-01 | -1.144214e-01 | -1.081250e-01 | -1.183666e-01 | -1.348015e-01 | -1.444428e-01 | -1.385527e-01 | -1.426251e-01 | â‹¯ |  5.650564e-02 |  2.639246e-02 |  1.109185e-01 |  6.937041e-02 |  8.791064e-02 |  1.055854e-01 |  1.084018e-01 |  7.893540e-02 |  7.554681e-02 |  1.749875e-01 |\n",
       "| 5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) | -1.475776e-02 | -2.136897e-02 | -2.183746e-02 | -1.868801e-02 | -1.880394e-02 | -1.911732e-02 | -2.733048e-02 | -3.572729e-02 | -3.150585e-02 | -2.717683e-02 | â‹¯ |  4.954903e-03 |  2.586807e-02 |  2.185875e-02 |  4.275850e-02 |  4.369622e-02 |  5.298591e-02 | -1.144091e-02 |  5.217841e-02 | -1.841314e-02 |  2.890525e-02 |\n",
       "| LV 6 |  1.137411e-02 |  1.100171e-02 |  2.021871e-02 |  2.901634e-02 |  1.232140e-02 |  1.205028e-02 |  9.152861e-03 |  1.060361e-02 | -6.413729e-04 |  6.352040e-03 | â‹¯ | -7.302503e-02 | -5.922457e-02 | -7.198972e-02 | -5.402172e-02 | -6.231897e-02 | -6.095302e-02 | -5.806830e-02 | -6.060223e-02 | -7.126057e-02 | -5.891329e-02 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                                       [,1]         \n",
       "LV 1                                                                    3.038562e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -3.281172e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -9.481710e-02\n",
       "4,Translation (GO:0006412)                                             -1.167416e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.475776e-02\n",
       "LV 6                                                                    1.137411e-02\n",
       "                                                                       [,2]         \n",
       "LV 1                                                                   -1.164697e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -3.819620e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -7.740678e-02\n",
       "4,Translation (GO:0006412)                                             -1.169561e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -2.136897e-02\n",
       "LV 6                                                                    1.100171e-02\n",
       "                                                                       [,3]         \n",
       "LV 1                                                                    1.499369e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -4.021169e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -9.260198e-02\n",
       "4,Translation (GO:0006412)                                             -1.241410e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -2.183746e-02\n",
       "LV 6                                                                    2.021871e-02\n",
       "                                                                       [,4]         \n",
       "LV 1                                                                    2.420048e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -4.102633e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -8.277387e-02\n",
       "4,Translation (GO:0006412)                                             -1.144214e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.868801e-02\n",
       "LV 6                                                                    2.901634e-02\n",
       "                                                                       [,5]         \n",
       "LV 1                                                                    1.076099e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -3.925897e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -7.081893e-02\n",
       "4,Translation (GO:0006412)                                             -1.081250e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.880394e-02\n",
       "LV 6                                                                    1.232140e-02\n",
       "                                                                       [,6]         \n",
       "LV 1                                                                    2.400729e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -3.957653e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -8.502113e-02\n",
       "4,Translation (GO:0006412)                                             -1.183666e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.911732e-02\n",
       "LV 6                                                                    1.205028e-02\n",
       "                                                                       [,7]         \n",
       "LV 1                                                                    2.450469e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -9.111154e-03\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -6.194781e-02\n",
       "4,Translation (GO:0006412)                                             -1.348015e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -2.733048e-02\n",
       "LV 6                                                                    9.152861e-03\n",
       "                                                                       [,8]         \n",
       "LV 1                                                                    2.529211e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -7.944571e-03\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -6.563719e-02\n",
       "4,Translation (GO:0006412)                                             -1.444428e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -3.572729e-02\n",
       "LV 6                                                                    1.060361e-02\n",
       "                                                                       [,9]         \n",
       "LV 1                                                                    2.492415e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -1.111526e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -6.688297e-02\n",
       "4,Translation (GO:0006412)                                             -1.385527e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -3.150585e-02\n",
       "LV 6                                                                   -6.413729e-04\n",
       "                                                                       [,10]        \n",
       "LV 1                                                                    2.411303e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -5.022869e-03\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -7.041583e-02\n",
       "4,Translation (GO:0006412)                                             -1.426251e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -2.717683e-02\n",
       "LV 6                                                                    6.352040e-03\n",
       "                                                                       [,11]\n",
       "LV 1                                                                   â‹¯    \n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         â‹¯    \n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          â‹¯    \n",
       "4,Translation (GO:0006412)                                             â‹¯    \n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) â‹¯    \n",
       "LV 6                                                                   â‹¯    \n",
       "                                                                       [,12]        \n",
       "LV 1                                                                   -2.252487e-14\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          3.445032e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -5.538873e-02\n",
       "4,Translation (GO:0006412)                                              5.650564e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  4.954903e-03\n",
       "LV 6                                                                   -7.302503e-02\n",
       "                                                                       [,13]        \n",
       "LV 1                                                                   -9.868426e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          9.405888e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -6.185108e-02\n",
       "4,Translation (GO:0006412)                                              2.639246e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  2.586807e-02\n",
       "LV 6                                                                   -5.922457e-02\n",
       "                                                                       [,14]        \n",
       "LV 1                                                                    7.459374e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          5.244139e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -3.681471e-02\n",
       "4,Translation (GO:0006412)                                              1.109185e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  2.185875e-02\n",
       "LV 6                                                                   -7.198972e-02\n",
       "                                                                       [,15]        \n",
       "LV 1                                                                   -1.238694e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          9.127825e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -3.960329e-02\n",
       "4,Translation (GO:0006412)                                              6.937041e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  4.275850e-02\n",
       "LV 6                                                                   -5.402172e-02\n",
       "                                                                       [,16]        \n",
       "LV 1                                                                   -5.471264e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          8.897180e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -6.117308e-02\n",
       "4,Translation (GO:0006412)                                              8.791064e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  4.369622e-02\n",
       "LV 6                                                                   -6.231897e-02\n",
       "                                                                       [,17]        \n",
       "LV 1                                                                    3.208637e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          6.534399e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -2.970421e-02\n",
       "4,Translation (GO:0006412)                                              1.055854e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  5.298591e-02\n",
       "LV 6                                                                   -6.095302e-02\n",
       "                                                                       [,18]        \n",
       "LV 1                                                                   -1.203242e-12\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          7.464888e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -8.491269e-02\n",
       "4,Translation (GO:0006412)                                              1.084018e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.144091e-02\n",
       "LV 6                                                                   -5.806830e-02\n",
       "                                                                       [,19]        \n",
       "LV 1                                                                   -9.133111e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          1.037498e-01\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -3.287660e-02\n",
       "4,Translation (GO:0006412)                                              7.893540e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  5.217841e-02\n",
       "LV 6                                                                   -6.060223e-02\n",
       "                                                                       [,20]        \n",
       "LV 1                                                                   -7.087071e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          4.380141e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -2.841848e-02\n",
       "4,Translation (GO:0006412)                                              7.554681e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.841314e-02\n",
       "LV 6                                                                   -7.126057e-02\n",
       "                                                                       [,21]        \n",
       "LV 1                                                                   -7.032934e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          9.150614e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -1.169766e-01\n",
       "4,Translation (GO:0006412)                                              1.749875e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  2.890525e-02\n",
       "LV 6                                                                   -5.891329e-02"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(gtex_plier$Z) <- paste0('LV', seq_len(ncol(gtex_plier$Z)))\n",
    "\n",
    "gtex_plier$summary <- gtex_plier$summary %>%\n",
    "    dplyr::rename(LV = `LV index`)  %>% \n",
    "    dplyr::mutate(LV = paste0('LV', LV))\n",
    "\n",
    "colnames(gtex_plier$B) <- samples\n",
    "\n",
    "saveRDS(gtex_plier, file = file.path(output_data_dir, \"PLIER.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "model_dir <- file.path(output_data_dir, \"PLIER\")\n",
    "dir.create(model_dir, showWarnings = FALSE, recursive = TRUE)\n",
    "\n",
    "B <- gtex_plier$B\n",
    "write.csv(B, file.path(model_dir, \"B.csv\"))\n",
    "\n",
    "Z <- gtex_plier$Z\n",
    "write.csv(Z, file.path(model_dir, \"Z.csv\"))\n",
    "\n",
    "summary <- gtex_plier$summary\n",
    "write.csv(summary, file.path(model_dir, \"summary.csv\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
