{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ca81bc7f",
   "metadata": {},
   "source": [
    "# recount2 model building with CLAMP and PLIER"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3ee54fb",
   "metadata": {},
   "source": [
    "ðŸ’¡ **Environment:** `clamp-analyses`  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56c6fd3c",
   "metadata": {},
   "source": [
    "This notebook builds latent variable models (CLAMP and PLIER) from recount2 RNAâ€‘seq RPKM data: it loads configuration and libraries, downloads and reads preprocessed recount2 data, maps Ensembl IDs to HGNC symbols and subsets genes and samples, stores the expression matrix in a bigstatsr FBM and applies filtering and zâ€‘scoring (preprocessCLAMPFBM), computes SVD (big_randomSVD) and infers CLAMP K, runs CLAMPbase and CLAMPfull with pathway priors and runs PLIER with matched priors for comparison, and saves model matrices (B, Z)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e69fa323",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "# CLAMP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be805cf1",
   "metadata": {},
   "source": [
    "## Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ec5663f0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recount2 CLAMP and PLIER analysis started at: 2025-11-25 10:04:14 \n"
     ]
    }
   ],
   "source": [
    "# Create a timestamp to track the start of the analysis\n",
    "start_time <- Sys.time()\n",
    "cat(\"recount2 CLAMP and PLIER analysis started at:\", format(start_time), \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b4ed5e6f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: â€˜dplyrâ€™\n",
      "\n",
      "\n",
      "The following objects are masked from â€˜package:data.tableâ€™:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from â€˜package:statsâ€™:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from â€˜package:baseâ€™:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: Matrix\n",
      "\n",
      "Loaded glmnet 4.1-10\n",
      "\n",
      "here() starts at /home/msubirana/Documents/pivlab/clamp-analyses\n",
      "\n",
      "Loading required package: RColorBrewer\n",
      "\n",
      "Loading required package: gplots\n",
      "\n",
      "\n",
      "Attaching package: â€˜gplotsâ€™\n",
      "\n",
      "\n",
      "The following object is masked from â€˜package:statsâ€™:\n",
      "\n",
      "    lowess\n",
      "\n",
      "\n",
      "Loading required package: pheatmap\n",
      "\n",
      "Loading required package: qvalue\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!requireNamespace(\"PLIER\", quietly = TRUE)) {\n",
    "    devtools::install_github(\"wgmao/PLIER\")\n",
    "}\n",
    "\n",
    "library(bigstatsr)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "library(rsvd)\n",
    "library(glmnet)\n",
    "library(Matrix)\n",
    "library(knitr)\n",
    "library(here)\n",
    "library(PLIER)\n",
    "library(CLAMP)\n",
    "\n",
    "source(here(\"config.R\"))\n",
    "\n",
    "set.seed(config$recount2$RANDOM_SVD_SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc858c79",
   "metadata": {},
   "source": [
    "## Output directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02f368b4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "output_data_dir <- config$recount2$DATASET_FOLDER\n",
    "dir.create(output_data_dir, showWarnings = FALSE, recursive = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7286f749",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "block_size <- config$GENERAL$CHUNK_SIZE\n",
    "N_CORES    <- config$recount2$N_CORES"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf137533",
   "metadata": {},
   "source": [
    "## Download recount2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "759882bd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Required files already exist. Skipping download and decompression.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "url      <- config$recount2$URL\n",
    "out_dir  <- here::here(\"data\", \"recount2\")\n",
    "zip_file <- file.path(out_dir, \"recount2.zip\")\n",
    "rds_file <- file.path(out_dir, \"recount2_PLIER_data\", \"recount_data_prep_PLIER.RDS\")\n",
    "rpkm_file <- file.path(out_dir, \"recount2_PLIER_data\", \"recount_rpkm.RDS\")\n",
    "\n",
    "if (!file.exists(rds_file) && !file.exists(rpkm_file)) {\n",
    "    if (!dir.exists(out_dir)) dir.create(out_dir, recursive = TRUE)\n",
    "    \n",
    "    system(paste('wget', url, '-O', zip_file))\n",
    "    unzip(zip_file, exdir = out_dir)\n",
    "} else {\n",
    "    message(\"Required files already exist. Skipping download and decompression.\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7680eee5",
   "metadata": {},
   "source": [
    "## Preprocess recount2 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41d0a2d0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Read preprocessed recount2 data\n",
    "data_mat <- readRDS(rds_file)\n",
    "\n",
    "meta <- list()\n",
    "meta$gene_symbols <- rownames(data_mat$rpkm.cm)\n",
    "meta$samples <- colnames(data_mat$rpkm.cm)\n",
    "\n",
    "rm(data_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c3dc65e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Read in recount2 file\n",
    "\n",
    "rpkm.df <- readRDS(rpkm_file)\n",
    "\n",
    "mart <- biomaRt::useDataset(\"hsapiens_gene_ensembl\",\n",
    "                            biomaRt::useMart(\"ensembl\"))\n",
    "\n",
    "genes <- unlist(lapply(strsplit(rpkm.df$ENSG, \"[.]\"), `[[`, 1))\n",
    "\n",
    "rpkm.df$ensembl_gene_id <- unlist(lapply(strsplit(rpkm.df$ENSG, \"[.]\"),\n",
    "                                         `[[`, 1))\n",
    "                                         \n",
    "gene.df <- biomaRt::getBM(filters = \"ensembl_gene_id\",\n",
    "                          attributes = c(\"ensembl_gene_id\", \"hgnc_symbol\"),\n",
    "                          values = genes,\n",
    "                          mart = mart)\n",
    "\n",
    "gene.df <- gene.df %>% dplyr::filter(complete.cases(.))\n",
    "\n",
    "rpkm.df <- dplyr::inner_join(gene.df, rpkm.df,\n",
    "                             by = \"ensembl_gene_id\",\n",
    "                             relationship = \"many-to-many\")\n",
    "\n",
    "rownames(rpkm.df) <- make.names(rpkm.df$hgnc_symbol, unique = TRUE)\n",
    "\n",
    "rpkm.df <- rpkm.df %>% dplyr::select(-c(ensembl_gene_id:ENSG))\n",
    "\n",
    "data_mat <- rpkm.df[meta$gene_symbols, meta$samples]\n",
    "\n",
    "rm(rpkm.df)\n",
    "\n",
    "data_mat <- as.matrix(as.data.table(data_mat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68af70e3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "âœ… Removed existing FBM file: /home/msubirana/Documents/pivlab/clamp-analyses/output/recount2/FBMrecount2.bk\n",
      "\n",
      "âœ… Removed existing FBM file: /home/msubirana/Documents/pivlab/clamp-analyses/output/recount2/FBMrecount2.bk_preproc.bk\n",
      "\n",
      "âœ… Removed existing FBM file: /home/msubirana/Documents/pivlab/clamp-analyses/output/recount2/FBMrecount2.bk_preproc_filtered.bk\n",
      "\n",
      "â„¹ï¸ Not found, skipping: /home/msubirana/Documents/pivlab/clamp-analyses/output/recount2/FBMrecount2.bk_preproc_filtered.rds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "fbm_files <- c(\n",
    "  file.path(output_data_dir, \"FBMrecount2.bk\"),\n",
    "  file.path(output_data_dir, \"FBMrecount2_preproc.bk\"),\n",
    "  file.path(output_data_dir, \"FBMrecount2_preproc_filtered.bk\"),\n",
    "  file.path(output_data_dir, \"FBMrecount2_preproc_filtered.rds\")\n",
    ")\n",
    "\n",
    "for (f in fbm_files) {\n",
    "  if (file.exists(f)) {\n",
    "    res <- unlink(f)\n",
    "    if (identical(res, 0L)) {\n",
    "      message(\"âœ… Removed existing FBM file: \", f)\n",
    "    } else {\n",
    "      warning(\"âš ï¸ Failed to remove: \", f)\n",
    "    }\n",
    "  } else {\n",
    "    message(\"â„¹ï¸ Not found, skipping: \", f)\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2fb30b5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Create the FBM\n",
    "n_genes <- length(meta$gene_symbols)\n",
    "n_samples <- length(meta$samples)\n",
    "\n",
    "fbm_file <- file.path(output_data_dir, \"FBMrecount2\")\n",
    "recount2FBM <- FBM(nrow = n_genes, ncol = n_samples, backingfile = fbm_file, create_bk = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62a459dd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Populate it with data\n",
    "block_size <- config$GENERAL$CHUNK_SIZE\n",
    "n_blocks <- ceiling(n_genes / block_size)\n",
    "\n",
    "for (i in 1:n_blocks) {\n",
    "  start_row <- (i-1) * block_size + 1\n",
    "  end_row <- min(i * block_size, nrow(data_mat))\n",
    "  recount2FBM[start_row:end_row, ] <- as.matrix(data_mat[start_row:end_row, ])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfe9908d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Applying log2 transformation\n",
      "\n",
      "Filling NAs with 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Preprocess and zâ€‘score FBM\n",
    "prep_recount2 <- preprocessCLAMPFBM(\n",
    "  fbm        = recount2FBM,\n",
    "  mean_cutoff= config$recount2$GENES_MEAN_CUTOFF,\n",
    "  var_cutoff = config$recount2$GENES_VAR_CUTOFF\n",
    ")\n",
    "\n",
    "recount2_fbm_filt <- prep_recount2$fbm_filtered\n",
    "recount2_rowStats <- prep_recount2$rowStats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee0ee831",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Applying Z-score transformation\n",
      "\n"
     ]
    }
   ],
   "source": [
    "zscoreCLAMPFBM(recount2_fbm_filt, recount2_rowStats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "samples <- meta$samples\n",
    "recount2_genes <- meta$gene_symbols[prep_recount2$kept_rows]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec6458f8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "saveRDS(samples, file = file.path(output_data_dir, \"recount2_samples.rds\"))\n",
    "saveRDS(recount2_genes, file = file.path(output_data_dir, \"recount2_genes.rds\"))\n",
    "saveRDS(recount2_fbm_filt, file = file.path(output_data_dir, \"recount2_fbm_filt.rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62186aa8",
   "metadata": {},
   "source": [
    "## SVD computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e3990c6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using SVD K = 5999\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!file.exists(file.path(output_data_dir, \"recount2_svdRes.rds\"))) {\n",
    "\n",
    "  g_fb <- nrow(recount2_fbm_filt)\n",
    "  samples_fb <- ncol(recount2_fbm_filt)\n",
    "  SVD_K_recount2 <- min(g_fb, samples_fb) - 1\n",
    "\n",
    "  message(\"Using SVD K = \", SVD_K_recount2)\n",
    "\n",
    "  if (N_CORES > 1) {\n",
    "    # disable BLAS parallelization when parallelizing\n",
    "    options(bigstatsr.check.parallel.blas = FALSE)\n",
    "    blas_nproc <- getOption(\"default.nproc.blas\")\n",
    "    options(default.nproc.blas = NULL)\n",
    "  }\n",
    "\n",
    "  recount2_svdRes <- big_randomSVD(\n",
    "    recount2_fbm_filt,\n",
    "    k = SVD_K_recount2,\n",
    "    ncores = N_CORES\n",
    "  )\n",
    "\n",
    "  saveRDS(recount2_svdRes, file = file.path(output_data_dir, \"recount2_svdRes.rds\"))\n",
    "\n",
    "  if (N_CORES > 1) {\n",
    "    # restore previous state\n",
    "    options(bigstatsr.check.parallel.blas = TRUE)\n",
    "    options(default.nproc.blas = blas_nproc)\n",
    "  }\n",
    "\n",
    "} else {\n",
    "  message(\"recount2_svdRes already exists, skipping SVD computation.\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dedb1810",
   "metadata": {},
   "source": [
    "## Estimate K for CLAMP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52d08c85",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Inferred CLAMP K = 832\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CLAMP_K_recount2 <- num.pc(list(d = recount2_svdRes$d)) * 2\n",
    "message(\"Inferred CLAMP K = \", CLAMP_K_recount2)\n",
    "\n",
    "saveRDS(CLAMP_K_recount2, file = file.path(output_data_dir, \"CLAMP_K_recount2.rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cdcb0f",
   "metadata": {},
   "source": [
    "## CLAMPbase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dc460fb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "****\n",
      "\n",
      "CLAMP k is set to 832\n",
      "\n",
      "L1 is set to 35.6888738143438\n",
      "\n",
      "L2 is set to 107.066621443032\n",
      "\n",
      "Progress 1 / 200 | Bdiff=0.459785, minCor=0.702589\n",
      "\n",
      "Progress 2 / 200 | Bdiff=0.035444, minCor=0.932119\n",
      "\n",
      "Progress 3 / 200 | Bdiff=0.013512, minCor=0.974439\n",
      "\n",
      "Progress 4 / 200 | Bdiff=0.008806, minCor=0.984172\n",
      "\n",
      "Progress 5 / 200 | Bdiff=0.006394, minCor=0.989010\n",
      "\n",
      "Progress 6 / 200 | Bdiff=0.004889, minCor=0.992503\n",
      "\n",
      "Progress 7 / 200 | Bdiff=0.003897, minCor=0.994383\n",
      "\n",
      "Progress 8 / 200 | Bdiff=0.003210, minCor=0.995265\n",
      "\n",
      "Progress 9 / 200 | Bdiff=0.002715, minCor=0.995975\n",
      "\n",
      "Progress 10 / 200 | Bdiff=0.002342, minCor=0.996665\n",
      "\n",
      "Progress 11 / 200 | Bdiff=0.002053, minCor=0.997100\n",
      "\n",
      "Progress 12 / 200 | Bdiff=0.001821, minCor=0.997500\n",
      "\n",
      "Progress 13 / 200 | Bdiff=0.001631, minCor=0.997689\n",
      "\n",
      "Progress 14 / 200 | Bdiff=0.001472, minCor=0.998018\n",
      "\n",
      "Progress 15 / 200 | Bdiff=0.001339, minCor=0.998277\n",
      "\n",
      "Progress 16 / 200 | Bdiff=0.001226, minCor=0.998497\n",
      "\n",
      "Progress 17 / 200 | Bdiff=0.001128, minCor=0.998655\n",
      "\n",
      "Progress 18 / 200 | Bdiff=0.001045, minCor=0.998747\n",
      "\n",
      "Progress 19 / 200 | Bdiff=0.000972, minCor=0.998842\n",
      "\n",
      "Progress 20 / 200 | Bdiff=0.019008, minCor=0.969234\n",
      "\n",
      "Progress 21 / 200 | Bdiff=0.008944, minCor=0.980212\n",
      "\n",
      "Progress 22 / 200 | Bdiff=0.005117, minCor=0.988626\n",
      "\n",
      "Progress 23 / 200 | Bdiff=0.003494, minCor=0.992052\n",
      "\n",
      "Progress 24 / 200 | Bdiff=0.002649, minCor=0.990658\n",
      "\n",
      "Progress 25 / 200 | Bdiff=0.002191, minCor=0.992999\n",
      "\n",
      "Progress 26 / 200 | Bdiff=0.001832, minCor=0.993843\n",
      "\n",
      "Progress 27 / 200 | Bdiff=0.001656, minCor=0.993010\n",
      "\n",
      "Progress 28 / 200 | Bdiff=0.001476, minCor=0.995448\n",
      "\n",
      "Progress 29 / 200 | Bdiff=0.001357, minCor=0.995571\n",
      "\n",
      "Progress 30 / 200 | Bdiff=0.001271, minCor=0.995470\n",
      "\n",
      "Progress 31 / 200 | Bdiff=0.001215, minCor=0.995245\n",
      "\n",
      "Progress 32 / 200 | Bdiff=0.001144, minCor=0.995755\n",
      "\n",
      "Progress 33 / 200 | Bdiff=0.001058, minCor=0.995796\n",
      "\n",
      "Progress 34 / 200 | Bdiff=0.000981, minCor=0.996850\n",
      "\n",
      "Progress 35 / 200 | Bdiff=0.000926, minCor=0.996588\n",
      "\n",
      "Progress 36 / 200 | Bdiff=0.000890, minCor=0.997795\n",
      "\n",
      "Progress 37 / 200 | Bdiff=0.000812, minCor=0.997989\n",
      "\n",
      "Progress 38 / 200 | Bdiff=0.000811, minCor=0.997946\n",
      "\n",
      "Progress 39 / 200 | Bdiff=0.000773, minCor=0.998071\n",
      "\n",
      "Progress 40 / 200 | Bdiff=0.000729, minCor=0.997905\n",
      "\n",
      "Progress 41 / 200 | Bdiff=0.000689, minCor=0.997780\n",
      "\n",
      "Progress 42 / 200 | Bdiff=0.000643, minCor=0.997846\n",
      "\n",
      "Progress 43 / 200 | Bdiff=0.000624, minCor=0.997863\n",
      "\n",
      "Progress 44 / 200 | Bdiff=0.000582, minCor=0.997828\n",
      "\n",
      "Progress 45 / 200 | Bdiff=0.000561, minCor=0.997736\n",
      "\n",
      "Progress 46 / 200 | Bdiff=0.000542, minCor=0.998097\n",
      "\n",
      "Progress 47 / 200 | Bdiff=0.000531, minCor=0.998284\n",
      "\n",
      "Progress 48 / 200 | Bdiff=0.000514, minCor=0.998329\n",
      "\n",
      "Progress 49 / 200 | Bdiff=0.000483, minCor=0.998589\n",
      "\n",
      "Converged at iteration= 49 | Bdiff=0.000483,  tol=0.000500     \n",
      "\n"
     ]
    }
   ],
   "source": [
    "recount2_baseRes <- CLAMPbase(\n",
    "  Y      = recount2_fbm_filt,\n",
    "  svdres = recount2_svdRes,\n",
    "  trace  = TRUE,\n",
    "  clamp_k = CLAMP_K_recount2\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "792f5f9f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 Ã— 832</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>LV1</th><th scope=col>LV2</th><th scope=col>LV3</th><th scope=col>LV4</th><th scope=col>LV5</th><th scope=col>LV6</th><th scope=col>LV7</th><th scope=col>LV8</th><th scope=col>LV9</th><th scope=col>LV10</th><th scope=col>â‹¯</th><th scope=col>LV823</th><th scope=col>LV824</th><th scope=col>LV825</th><th scope=col>LV826</th><th scope=col>LV827</th><th scope=col>LV828</th><th scope=col>LV829</th><th scope=col>LV830</th><th scope=col>LV831</th><th scope=col>LV832</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>â‹¯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GAS6</th><td>1.656654</td><td>0</td><td>0</td><td>0</td><td>0.0000000</td><td>0.8503073</td><td>0.2803847</td><td>0</td><td>0.6404271</td><td>0.2422511</td><td>â‹¯</td><td>0</td><td>0.2229221</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>MMP14</th><td>1.490046</td><td>0</td><td>0</td><td>0</td><td>0.0000000</td><td>0.4899336</td><td>0.0000000</td><td>0</td><td>0.4989182</td><td>0.0000000</td><td>â‹¯</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>DSP</th><td>1.642300</td><td>0</td><td>0</td><td>0</td><td>0.0000000</td><td>0.4149664</td><td>0.4115462</td><td>0</td><td>0.2766311</td><td>0.0000000</td><td>â‹¯</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>MARCKSL1</th><td>2.899502</td><td>0</td><td>0</td><td>0</td><td>0.3422124</td><td>0.0000000</td><td>0.0000000</td><td>0</td><td>0.0000000</td><td>0.0000000</td><td>â‹¯</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>SPARC</th><td>1.991356</td><td>0</td><td>0</td><td>0</td><td>0.0000000</td><td>0.5767266</td><td>0.0000000</td><td>0</td><td>0.4730226</td><td>0.0000000</td><td>â‹¯</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>CTSD</th><td>2.906868</td><td>0</td><td>0</td><td>0</td><td>0.0000000</td><td>0.0000000</td><td>0.0000000</td><td>0</td><td>0.0000000</td><td>0.0000000</td><td>â‹¯</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 Ã— 832\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & LV1 & LV2 & LV3 & LV4 & LV5 & LV6 & LV7 & LV8 & LV9 & LV10 & â‹¯ & LV823 & LV824 & LV825 & LV826 & LV827 & LV828 & LV829 & LV830 & LV831 & LV832\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & â‹¯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tGAS6 & 1.656654 & 0 & 0 & 0 & 0.0000000 & 0.8503073 & 0.2803847 & 0 & 0.6404271 & 0.2422511 & â‹¯ & 0 & 0.2229221 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tMMP14 & 1.490046 & 0 & 0 & 0 & 0.0000000 & 0.4899336 & 0.0000000 & 0 & 0.4989182 & 0.0000000 & â‹¯ & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tDSP & 1.642300 & 0 & 0 & 0 & 0.0000000 & 0.4149664 & 0.4115462 & 0 & 0.2766311 & 0.0000000 & â‹¯ & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tMARCKSL1 & 2.899502 & 0 & 0 & 0 & 0.3422124 & 0.0000000 & 0.0000000 & 0 & 0.0000000 & 0.0000000 & â‹¯ & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tSPARC & 1.991356 & 0 & 0 & 0 & 0.0000000 & 0.5767266 & 0.0000000 & 0 & 0.4730226 & 0.0000000 & â‹¯ & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tCTSD & 2.906868 & 0 & 0 & 0 & 0.0000000 & 0.0000000 & 0.0000000 & 0 & 0.0000000 & 0.0000000 & â‹¯ & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 Ã— 832\n",
       "\n",
       "| <!--/--> | LV1 &lt;dbl&gt; | LV2 &lt;dbl&gt; | LV3 &lt;dbl&gt; | LV4 &lt;dbl&gt; | LV5 &lt;dbl&gt; | LV6 &lt;dbl&gt; | LV7 &lt;dbl&gt; | LV8 &lt;dbl&gt; | LV9 &lt;dbl&gt; | LV10 &lt;dbl&gt; | â‹¯ â‹¯ | LV823 &lt;dbl&gt; | LV824 &lt;dbl&gt; | LV825 &lt;dbl&gt; | LV826 &lt;dbl&gt; | LV827 &lt;dbl&gt; | LV828 &lt;dbl&gt; | LV829 &lt;dbl&gt; | LV830 &lt;dbl&gt; | LV831 &lt;dbl&gt; | LV832 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| GAS6 | 1.656654 | 0 | 0 | 0 | 0.0000000 | 0.8503073 | 0.2803847 | 0 | 0.6404271 | 0.2422511 | â‹¯ | 0 | 0.2229221 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| MMP14 | 1.490046 | 0 | 0 | 0 | 0.0000000 | 0.4899336 | 0.0000000 | 0 | 0.4989182 | 0.0000000 | â‹¯ | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| DSP | 1.642300 | 0 | 0 | 0 | 0.0000000 | 0.4149664 | 0.4115462 | 0 | 0.2766311 | 0.0000000 | â‹¯ | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| MARCKSL1 | 2.899502 | 0 | 0 | 0 | 0.3422124 | 0.0000000 | 0.0000000 | 0 | 0.0000000 | 0.0000000 | â‹¯ | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| SPARC | 1.991356 | 0 | 0 | 0 | 0.0000000 | 0.5767266 | 0.0000000 | 0 | 0.4730226 | 0.0000000 | â‹¯ | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| CTSD | 2.906868 | 0 | 0 | 0 | 0.0000000 | 0.0000000 | 0.0000000 | 0 | 0.0000000 | 0.0000000 | â‹¯ | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "         LV1      LV2 LV3 LV4 LV5       LV6       LV7       LV8 LV9      \n",
       "GAS6     1.656654 0   0   0   0.0000000 0.8503073 0.2803847 0   0.6404271\n",
       "MMP14    1.490046 0   0   0   0.0000000 0.4899336 0.0000000 0   0.4989182\n",
       "DSP      1.642300 0   0   0   0.0000000 0.4149664 0.4115462 0   0.2766311\n",
       "MARCKSL1 2.899502 0   0   0   0.3422124 0.0000000 0.0000000 0   0.0000000\n",
       "SPARC    1.991356 0   0   0   0.0000000 0.5767266 0.0000000 0   0.4730226\n",
       "CTSD     2.906868 0   0   0   0.0000000 0.0000000 0.0000000 0   0.0000000\n",
       "         LV10      â‹¯ LV823 LV824     LV825 LV826 LV827 LV828 LV829 LV830 LV831\n",
       "GAS6     0.2422511 â‹¯ 0     0.2229221 0     0     0     0     0     0     0    \n",
       "MMP14    0.0000000 â‹¯ 0     0.0000000 0     0     0     0     0     0     0    \n",
       "DSP      0.0000000 â‹¯ 0     0.0000000 0     0     0     0     0     0     0    \n",
       "MARCKSL1 0.0000000 â‹¯ 0     0.0000000 0     0     0     0     0     0     0    \n",
       "SPARC    0.0000000 â‹¯ 0     0.0000000 0     0     0     0     0     0     0    \n",
       "CTSD     0.0000000 â‹¯ 0     0.0000000 0     0     0     0     0     0     0    \n",
       "         LV832\n",
       "GAS6     0    \n",
       "MMP14    0    \n",
       "DSP      0    \n",
       "MARCKSL1 0    \n",
       "SPARC    0    \n",
       "CTSD     0    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "recount2_baseRes$Z <- data.frame(recount2_baseRes$Z)\n",
    "rownames(recount2_baseRes$Z) <- recount2_genes\n",
    "\n",
    "recount2_baseRes$B <- data.frame(recount2_baseRes$B)\n",
    "colnames(recount2_baseRes$B) <- samples\n",
    "\n",
    "saveRDS(recount2_baseRes, file = file.path(output_data_dir, \"CLAMPbase.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "model_dir <- file.path(output_data_dir, \"CLAMPbase\")\n",
    "dir.create(model_dir, showWarnings = FALSE, recursive = TRUE)\n",
    "\n",
    "B <- recount2_baseRes$B\n",
    "write.csv(B, file.path(model_dir, \"B.csv\"))\n",
    "\n",
    "Z <- recount2_baseRes$Z\n",
    "write.csv(Z, file.path(model_dir, \"Z.csv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ee267bb",
   "metadata": {},
   "source": [
    "## Prepare pathway priors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "960031ac",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Auto-detected name: KEGG_2021_Human\n",
      "\n",
      "Using cached file for KEGG_2021_Human\n",
      "\n",
      "Auto-detected name: GO_Biological_Process_2025\n",
      "\n",
      "Using cached file for GO_Biological_Process_2025\n",
      "\n",
      "Auto-detected name: GTEx_Tissues_V8_2023\n",
      "\n",
      "Using cached file for GTEx_Tissues_V8_2023\n",
      "\n",
      "There are 5844 genes in the intersection between data and prior\n",
      "\n",
      "Removing 3296 pathways\n",
      "\n",
      "Inverting...\n",
      "\n",
      "done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "recount2_gmtList <- list(\n",
    "  KEGG = getGMT(\"https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName=KEGG_2021_Human\"),\n",
    "  BP = getGMT(\"https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName=GO_Biological_Process_2025\"),\n",
    "  GTEx_Tissues = getGMT(\"https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName=GTEx_Tissues_V8_2023\")\n",
    ")\n",
    "\n",
    "# prefix each geneâ€set name with its library to guarantee uniqueness\n",
    "for(lib in names(recount2_gmtList)) {\n",
    "  names(recount2_gmtList[[lib]]) <- paste0(lib, \"_\", names(recount2_gmtList[[lib]]))\n",
    "}\n",
    "\n",
    "recount2_pathMat <- gmtListToSparseMat(recount2_gmtList)\n",
    "recount2_matched <- getMatchedPathwayMat(recount2_pathMat, recount2_genes)\n",
    "recount2_chatObj <- getChat(recount2_matched)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6526304d",
   "metadata": {},
   "source": [
    "## CLAMPfull"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5673ed3a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "** CLAMPfull **\n",
      "\n",
      "using provided CLAMPbase result\n",
      "\n",
      "CLAMP k is set to 832\n",
      "\n",
      "L1=35.6888738143438; L2=107.066621443032\n",
      "\n",
      "Progress 1 / 30 | Bdiff=0.000458\n",
      "\n",
      "Progress 2 / 30 | Bdiff=0.007594\n",
      "\n",
      "Progress 3 / 30 | Bdiff=0.007064\n",
      "\n",
      "Estimated total runtime: ~6.6 min\n",
      "\n",
      "Progress 4 / 30 | Bdiff=0.007791\n",
      "\n",
      "Progress 5 / 30 | Bdiff=0.009431\n",
      "\n",
      "Progress 6 / 30 | Bdiff=0.010059\n",
      "\n",
      "Progress 7 / 30 | Bdiff=0.011233\n",
      "\n",
      "Progress 8 / 30 | Bdiff=0.014777\n",
      "\n",
      "Progress 9 / 30 | Bdiff=0.012048\n",
      "\n",
      "Progress 10 / 30 | Bdiff=0.011607\n",
      "\n",
      "Progress 11 / 30 | Bdiff=0.018083\n",
      "\n",
      "Progress 12 / 30 | Bdiff=0.011340\n",
      "\n",
      "Progress 13 / 30 | Bdiff=0.010272\n",
      "\n",
      "Progress 14 / 30 | Bdiff=0.019521\n",
      "\n",
      "Progress 15 / 30 | Bdiff=0.011078\n",
      "\n",
      "Progress 16 / 30 | Bdiff=0.009438\n",
      "\n",
      "Progress 17 / 30 | Bdiff=0.018238\n",
      "\n",
      "Progress 18 / 30 | Bdiff=0.010508\n",
      "\n",
      "Progress 19 / 30 | Bdiff=0.009156\n",
      "\n",
      "Progress 20 / 30 | Bdiff=0.019579\n",
      "\n",
      "Progress 21 / 30 | Bdiff=0.008802\n",
      "\n",
      "Progress 22 / 30 | Bdiff=0.007460\n",
      "\n",
      "Progress 23 / 30 | Bdiff=0.017314\n",
      "\n",
      "Progress 24 / 30 | Bdiff=0.008235\n",
      "\n",
      "Progress 25 / 30 | Bdiff=0.007214\n",
      "\n",
      "Progress 26 / 30 | Bdiff=0.016458\n",
      "\n",
      "Progress 27 / 30 | Bdiff=0.007932\n",
      "\n",
      "Progress 28 / 30 | Bdiff=0.007151\n",
      "\n",
      "Progress 29 / 30 | Bdiff=0.017057\n",
      "\n",
      "Progress 30 / 30 | Bdiff=0.007773\n",
      "\n",
      "Updating Z for CV\n",
      "\n",
      "crossValidation\n",
      "\n",
      "There are 546 LVs with AUC>0.70\n",
      "\n",
      "There are 228 LVs with AUC>0.90\n",
      "\n"
     ]
    }
   ],
   "source": [
    "recount2_fullRes <- CLAMPfull(\n",
    "    Y = recount2_fbm_filt,\n",
    "    svdres = recount2_svdRes,\n",
    "    priorMat = recount2_matched,\n",
    "    clamp.base.result = recount2_baseRes,\n",
    "    use_cpp = TRUE,\n",
    "    trace = TRUE,\n",
    "    clamp_k = CLAMP_K_recount2\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2a0236d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 Ã— 832</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>LV1</th><th scope=col>LV2</th><th scope=col>LV3</th><th scope=col>LV4</th><th scope=col>LV5</th><th scope=col>LV6</th><th scope=col>LV7</th><th scope=col>LV8</th><th scope=col>LV9</th><th scope=col>LV10</th><th scope=col>â‹¯</th><th scope=col>LV823</th><th scope=col>LV824</th><th scope=col>LV825</th><th scope=col>LV826</th><th scope=col>LV827</th><th scope=col>LV828</th><th scope=col>LV829</th><th scope=col>LV830</th><th scope=col>LV831</th><th scope=col>LV832</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>â‹¯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GAS6</th><td>1.656654</td><td>0</td><td>0</td><td>0</td><td>0.0000000</td><td>0.8503073</td><td>0.2803847</td><td>0</td><td>0.6404271</td><td>0.2422511</td><td>â‹¯</td><td>0</td><td>0.2229221</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>MMP14</th><td>1.490046</td><td>0</td><td>0</td><td>0</td><td>0.0000000</td><td>0.4899336</td><td>0.0000000</td><td>0</td><td>0.4989182</td><td>0.0000000</td><td>â‹¯</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>DSP</th><td>1.642300</td><td>0</td><td>0</td><td>0</td><td>0.0000000</td><td>0.4149664</td><td>0.4115462</td><td>0</td><td>0.2766311</td><td>0.0000000</td><td>â‹¯</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>MARCKSL1</th><td>2.899502</td><td>0</td><td>0</td><td>0</td><td>0.3422124</td><td>0.0000000</td><td>0.0000000</td><td>0</td><td>0.0000000</td><td>0.0000000</td><td>â‹¯</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>SPARC</th><td>1.991356</td><td>0</td><td>0</td><td>0</td><td>0.0000000</td><td>0.5767266</td><td>0.0000000</td><td>0</td><td>0.4730226</td><td>0.0000000</td><td>â‹¯</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>CTSD</th><td>2.906868</td><td>0</td><td>0</td><td>0</td><td>0.0000000</td><td>0.0000000</td><td>0.0000000</td><td>0</td><td>0.0000000</td><td>0.0000000</td><td>â‹¯</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 Ã— 832\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & LV1 & LV2 & LV3 & LV4 & LV5 & LV6 & LV7 & LV8 & LV9 & LV10 & â‹¯ & LV823 & LV824 & LV825 & LV826 & LV827 & LV828 & LV829 & LV830 & LV831 & LV832\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & â‹¯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tGAS6 & 1.656654 & 0 & 0 & 0 & 0.0000000 & 0.8503073 & 0.2803847 & 0 & 0.6404271 & 0.2422511 & â‹¯ & 0 & 0.2229221 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tMMP14 & 1.490046 & 0 & 0 & 0 & 0.0000000 & 0.4899336 & 0.0000000 & 0 & 0.4989182 & 0.0000000 & â‹¯ & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tDSP & 1.642300 & 0 & 0 & 0 & 0.0000000 & 0.4149664 & 0.4115462 & 0 & 0.2766311 & 0.0000000 & â‹¯ & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tMARCKSL1 & 2.899502 & 0 & 0 & 0 & 0.3422124 & 0.0000000 & 0.0000000 & 0 & 0.0000000 & 0.0000000 & â‹¯ & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tSPARC & 1.991356 & 0 & 0 & 0 & 0.0000000 & 0.5767266 & 0.0000000 & 0 & 0.4730226 & 0.0000000 & â‹¯ & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\tCTSD & 2.906868 & 0 & 0 & 0 & 0.0000000 & 0.0000000 & 0.0000000 & 0 & 0.0000000 & 0.0000000 & â‹¯ & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 Ã— 832\n",
       "\n",
       "| <!--/--> | LV1 &lt;dbl&gt; | LV2 &lt;dbl&gt; | LV3 &lt;dbl&gt; | LV4 &lt;dbl&gt; | LV5 &lt;dbl&gt; | LV6 &lt;dbl&gt; | LV7 &lt;dbl&gt; | LV8 &lt;dbl&gt; | LV9 &lt;dbl&gt; | LV10 &lt;dbl&gt; | â‹¯ â‹¯ | LV823 &lt;dbl&gt; | LV824 &lt;dbl&gt; | LV825 &lt;dbl&gt; | LV826 &lt;dbl&gt; | LV827 &lt;dbl&gt; | LV828 &lt;dbl&gt; | LV829 &lt;dbl&gt; | LV830 &lt;dbl&gt; | LV831 &lt;dbl&gt; | LV832 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| GAS6 | 1.656654 | 0 | 0 | 0 | 0.0000000 | 0.8503073 | 0.2803847 | 0 | 0.6404271 | 0.2422511 | â‹¯ | 0 | 0.2229221 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| MMP14 | 1.490046 | 0 | 0 | 0 | 0.0000000 | 0.4899336 | 0.0000000 | 0 | 0.4989182 | 0.0000000 | â‹¯ | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| DSP | 1.642300 | 0 | 0 | 0 | 0.0000000 | 0.4149664 | 0.4115462 | 0 | 0.2766311 | 0.0000000 | â‹¯ | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| MARCKSL1 | 2.899502 | 0 | 0 | 0 | 0.3422124 | 0.0000000 | 0.0000000 | 0 | 0.0000000 | 0.0000000 | â‹¯ | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| SPARC | 1.991356 | 0 | 0 | 0 | 0.0000000 | 0.5767266 | 0.0000000 | 0 | 0.4730226 | 0.0000000 | â‹¯ | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| CTSD | 2.906868 | 0 | 0 | 0 | 0.0000000 | 0.0000000 | 0.0000000 | 0 | 0.0000000 | 0.0000000 | â‹¯ | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "         LV1      LV2 LV3 LV4 LV5       LV6       LV7       LV8 LV9      \n",
       "GAS6     1.656654 0   0   0   0.0000000 0.8503073 0.2803847 0   0.6404271\n",
       "MMP14    1.490046 0   0   0   0.0000000 0.4899336 0.0000000 0   0.4989182\n",
       "DSP      1.642300 0   0   0   0.0000000 0.4149664 0.4115462 0   0.2766311\n",
       "MARCKSL1 2.899502 0   0   0   0.3422124 0.0000000 0.0000000 0   0.0000000\n",
       "SPARC    1.991356 0   0   0   0.0000000 0.5767266 0.0000000 0   0.4730226\n",
       "CTSD     2.906868 0   0   0   0.0000000 0.0000000 0.0000000 0   0.0000000\n",
       "         LV10      â‹¯ LV823 LV824     LV825 LV826 LV827 LV828 LV829 LV830 LV831\n",
       "GAS6     0.2422511 â‹¯ 0     0.2229221 0     0     0     0     0     0     0    \n",
       "MMP14    0.0000000 â‹¯ 0     0.0000000 0     0     0     0     0     0     0    \n",
       "DSP      0.0000000 â‹¯ 0     0.0000000 0     0     0     0     0     0     0    \n",
       "MARCKSL1 0.0000000 â‹¯ 0     0.0000000 0     0     0     0     0     0     0    \n",
       "SPARC    0.0000000 â‹¯ 0     0.0000000 0     0     0     0     0     0     0    \n",
       "CTSD     0.0000000 â‹¯ 0     0.0000000 0     0     0     0     0     0     0    \n",
       "         LV832\n",
       "GAS6     0    \n",
       "MMP14    0    \n",
       "DSP      0    \n",
       "MARCKSL1 0    \n",
       "SPARC    0    \n",
       "CTSD     0    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 Ã— 37032</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>SRP000599.SRR013549</th><th scope=col>SRP000599.SRR013550</th><th scope=col>SRP000599.SRR013551</th><th scope=col>SRP000599.SRR013552</th><th scope=col>SRP000599.SRR013553</th><th scope=col>SRP000599.SRR013554</th><th scope=col>SRP000599.SRR013555</th><th scope=col>SRP000599.SRR013556</th><th scope=col>SRP000599.SRR013557</th><th scope=col>SRP000599.SRR013558</th><th scope=col>â‹¯</th><th scope=col>SRP035599.SRR1139372</th><th scope=col>SRP035599.SRR1139393</th><th scope=col>SRP035599.SRR1139388</th><th scope=col>SRP035599.SRR1139378</th><th scope=col>SRP035599.SRR1139399</th><th scope=col>SRP035599.SRR1139386</th><th scope=col>SRP035599.SRR1139375</th><th scope=col>SRP035599.SRR1139382</th><th scope=col>SRP035599.SRR1139356</th><th scope=col>SRP035599.SRR1139370</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>â‹¯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>LV1</th><td>-0.867384901</td><td>-0.879844435</td><td>-0.873107420</td><td>-0.87148566</td><td>-0.8752239706</td><td>-0.875151653</td><td>-0.82379528</td><td>-8.266395e-01</td><td>-0.82067562</td><td>-0.82243867</td><td>â‹¯</td><td>-0.52053021</td><td>-0.51392282</td><td>-0.5257386</td><td>-0.5168343</td><td>-0.5285092</td><td>-0.5401691</td><td>-0.5359329</td><td>-0.5174164</td><td>-0.5040608</td><td>-0.5383226</td></tr>\n",
       "\t<tr><th scope=row>LV2</th><td>-0.133615298</td><td>-0.127031401</td><td>-0.128278993</td><td>-0.13233526</td><td>-0.1323391474</td><td>-0.131399397</td><td>-0.16725299</td><td>-1.608620e-01</td><td>-0.16321947</td><td>-0.16428624</td><td>â‹¯</td><td> 0.20624147</td><td> 0.08443934</td><td> 0.1507070</td><td> 0.1775285</td><td> 0.1510824</td><td> 0.1872756</td><td> 0.1662790</td><td> 0.1755454</td><td> 0.2382462</td><td> 0.1948014</td></tr>\n",
       "\t<tr><th scope=row>LV3</th><td>-0.033435328</td><td>-0.025611776</td><td>-0.023630533</td><td>-0.01513276</td><td>-0.0113543157</td><td>-0.024335924</td><td>-0.03823130</td><td>-3.274484e-02</td><td>-0.04298509</td><td>-0.04094039</td><td>â‹¯</td><td>-0.17209869</td><td>-0.16298852</td><td>-0.1731993</td><td>-0.1882488</td><td>-0.1720055</td><td>-0.1600504</td><td>-0.1761666</td><td>-0.1949913</td><td>-0.1505919</td><td>-0.1967838</td></tr>\n",
       "\t<tr><th scope=row>LV4</th><td>-0.243989908</td><td>-0.235868650</td><td>-0.244113406</td><td>-0.24443976</td><td>-0.2327881755</td><td>-0.236407836</td><td>-0.21912488</td><td>-2.010877e-01</td><td>-0.22754310</td><td>-0.21997184</td><td>â‹¯</td><td>-0.09571376</td><td>-0.14086539</td><td>-0.2011386</td><td>-0.1156240</td><td>-0.2145210</td><td>-0.1652276</td><td>-0.2586825</td><td>-0.1696389</td><td>-0.1651986</td><td>-0.3765665</td></tr>\n",
       "\t<tr><th scope=row>LV5</th><td>-0.019908724</td><td>-0.018702933</td><td>-0.008825051</td><td>-0.02083118</td><td>-0.0228384570</td><td>-0.017638770</td><td>-0.04534587</td><td>-4.036835e-02</td><td>-0.03258473</td><td>-0.03866460</td><td>â‹¯</td><td>-0.14597781</td><td>-0.14888982</td><td>-0.1407341</td><td>-0.1453838</td><td>-0.1405794</td><td>-0.1426756</td><td>-0.1224374</td><td>-0.1401182</td><td>-0.1452800</td><td>-0.1319524</td></tr>\n",
       "\t<tr><th scope=row>LV6</th><td>-0.006614441</td><td>-0.007498693</td><td>-0.008060589</td><td>-0.01517550</td><td> 0.0009615722</td><td>-0.008850218</td><td>-0.01707047</td><td>-7.413137e-05</td><td>-0.02293963</td><td>-0.01355327</td><td>â‹¯</td><td>-0.24519268</td><td>-0.24951182</td><td>-0.2527177</td><td>-0.2627610</td><td>-0.2605949</td><td>-0.2633666</td><td>-0.2431627</td><td>-0.2478865</td><td>-0.2298372</td><td>-0.2663158</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 Ã— 37032\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & SRP000599.SRR013549 & SRP000599.SRR013550 & SRP000599.SRR013551 & SRP000599.SRR013552 & SRP000599.SRR013553 & SRP000599.SRR013554 & SRP000599.SRR013555 & SRP000599.SRR013556 & SRP000599.SRR013557 & SRP000599.SRR013558 & â‹¯ & SRP035599.SRR1139372 & SRP035599.SRR1139393 & SRP035599.SRR1139388 & SRP035599.SRR1139378 & SRP035599.SRR1139399 & SRP035599.SRR1139386 & SRP035599.SRR1139375 & SRP035599.SRR1139382 & SRP035599.SRR1139356 & SRP035599.SRR1139370\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & â‹¯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tLV1 & -0.867384901 & -0.879844435 & -0.873107420 & -0.87148566 & -0.8752239706 & -0.875151653 & -0.82379528 & -8.266395e-01 & -0.82067562 & -0.82243867 & â‹¯ & -0.52053021 & -0.51392282 & -0.5257386 & -0.5168343 & -0.5285092 & -0.5401691 & -0.5359329 & -0.5174164 & -0.5040608 & -0.5383226\\\\\n",
       "\tLV2 & -0.133615298 & -0.127031401 & -0.128278993 & -0.13233526 & -0.1323391474 & -0.131399397 & -0.16725299 & -1.608620e-01 & -0.16321947 & -0.16428624 & â‹¯ &  0.20624147 &  0.08443934 &  0.1507070 &  0.1775285 &  0.1510824 &  0.1872756 &  0.1662790 &  0.1755454 &  0.2382462 &  0.1948014\\\\\n",
       "\tLV3 & -0.033435328 & -0.025611776 & -0.023630533 & -0.01513276 & -0.0113543157 & -0.024335924 & -0.03823130 & -3.274484e-02 & -0.04298509 & -0.04094039 & â‹¯ & -0.17209869 & -0.16298852 & -0.1731993 & -0.1882488 & -0.1720055 & -0.1600504 & -0.1761666 & -0.1949913 & -0.1505919 & -0.1967838\\\\\n",
       "\tLV4 & -0.243989908 & -0.235868650 & -0.244113406 & -0.24443976 & -0.2327881755 & -0.236407836 & -0.21912488 & -2.010877e-01 & -0.22754310 & -0.21997184 & â‹¯ & -0.09571376 & -0.14086539 & -0.2011386 & -0.1156240 & -0.2145210 & -0.1652276 & -0.2586825 & -0.1696389 & -0.1651986 & -0.3765665\\\\\n",
       "\tLV5 & -0.019908724 & -0.018702933 & -0.008825051 & -0.02083118 & -0.0228384570 & -0.017638770 & -0.04534587 & -4.036835e-02 & -0.03258473 & -0.03866460 & â‹¯ & -0.14597781 & -0.14888982 & -0.1407341 & -0.1453838 & -0.1405794 & -0.1426756 & -0.1224374 & -0.1401182 & -0.1452800 & -0.1319524\\\\\n",
       "\tLV6 & -0.006614441 & -0.007498693 & -0.008060589 & -0.01517550 &  0.0009615722 & -0.008850218 & -0.01707047 & -7.413137e-05 & -0.02293963 & -0.01355327 & â‹¯ & -0.24519268 & -0.24951182 & -0.2527177 & -0.2627610 & -0.2605949 & -0.2633666 & -0.2431627 & -0.2478865 & -0.2298372 & -0.2663158\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 Ã— 37032\n",
       "\n",
       "| <!--/--> | SRP000599.SRR013549 &lt;dbl&gt; | SRP000599.SRR013550 &lt;dbl&gt; | SRP000599.SRR013551 &lt;dbl&gt; | SRP000599.SRR013552 &lt;dbl&gt; | SRP000599.SRR013553 &lt;dbl&gt; | SRP000599.SRR013554 &lt;dbl&gt; | SRP000599.SRR013555 &lt;dbl&gt; | SRP000599.SRR013556 &lt;dbl&gt; | SRP000599.SRR013557 &lt;dbl&gt; | SRP000599.SRR013558 &lt;dbl&gt; | â‹¯ â‹¯ | SRP035599.SRR1139372 &lt;dbl&gt; | SRP035599.SRR1139393 &lt;dbl&gt; | SRP035599.SRR1139388 &lt;dbl&gt; | SRP035599.SRR1139378 &lt;dbl&gt; | SRP035599.SRR1139399 &lt;dbl&gt; | SRP035599.SRR1139386 &lt;dbl&gt; | SRP035599.SRR1139375 &lt;dbl&gt; | SRP035599.SRR1139382 &lt;dbl&gt; | SRP035599.SRR1139356 &lt;dbl&gt; | SRP035599.SRR1139370 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| LV1 | -0.867384901 | -0.879844435 | -0.873107420 | -0.87148566 | -0.8752239706 | -0.875151653 | -0.82379528 | -8.266395e-01 | -0.82067562 | -0.82243867 | â‹¯ | -0.52053021 | -0.51392282 | -0.5257386 | -0.5168343 | -0.5285092 | -0.5401691 | -0.5359329 | -0.5174164 | -0.5040608 | -0.5383226 |\n",
       "| LV2 | -0.133615298 | -0.127031401 | -0.128278993 | -0.13233526 | -0.1323391474 | -0.131399397 | -0.16725299 | -1.608620e-01 | -0.16321947 | -0.16428624 | â‹¯ |  0.20624147 |  0.08443934 |  0.1507070 |  0.1775285 |  0.1510824 |  0.1872756 |  0.1662790 |  0.1755454 |  0.2382462 |  0.1948014 |\n",
       "| LV3 | -0.033435328 | -0.025611776 | -0.023630533 | -0.01513276 | -0.0113543157 | -0.024335924 | -0.03823130 | -3.274484e-02 | -0.04298509 | -0.04094039 | â‹¯ | -0.17209869 | -0.16298852 | -0.1731993 | -0.1882488 | -0.1720055 | -0.1600504 | -0.1761666 | -0.1949913 | -0.1505919 | -0.1967838 |\n",
       "| LV4 | -0.243989908 | -0.235868650 | -0.244113406 | -0.24443976 | -0.2327881755 | -0.236407836 | -0.21912488 | -2.010877e-01 | -0.22754310 | -0.21997184 | â‹¯ | -0.09571376 | -0.14086539 | -0.2011386 | -0.1156240 | -0.2145210 | -0.1652276 | -0.2586825 | -0.1696389 | -0.1651986 | -0.3765665 |\n",
       "| LV5 | -0.019908724 | -0.018702933 | -0.008825051 | -0.02083118 | -0.0228384570 | -0.017638770 | -0.04534587 | -4.036835e-02 | -0.03258473 | -0.03866460 | â‹¯ | -0.14597781 | -0.14888982 | -0.1407341 | -0.1453838 | -0.1405794 | -0.1426756 | -0.1224374 | -0.1401182 | -0.1452800 | -0.1319524 |\n",
       "| LV6 | -0.006614441 | -0.007498693 | -0.008060589 | -0.01517550 |  0.0009615722 | -0.008850218 | -0.01707047 | -7.413137e-05 | -0.02293963 | -0.01355327 | â‹¯ | -0.24519268 | -0.24951182 | -0.2527177 | -0.2627610 | -0.2605949 | -0.2633666 | -0.2431627 | -0.2478865 | -0.2298372 | -0.2663158 |\n",
       "\n"
      ],
      "text/plain": [
       "    SRP000599.SRR013549 SRP000599.SRR013550 SRP000599.SRR013551\n",
       "LV1 -0.867384901        -0.879844435        -0.873107420       \n",
       "LV2 -0.133615298        -0.127031401        -0.128278993       \n",
       "LV3 -0.033435328        -0.025611776        -0.023630533       \n",
       "LV4 -0.243989908        -0.235868650        -0.244113406       \n",
       "LV5 -0.019908724        -0.018702933        -0.008825051       \n",
       "LV6 -0.006614441        -0.007498693        -0.008060589       \n",
       "    SRP000599.SRR013552 SRP000599.SRR013553 SRP000599.SRR013554\n",
       "LV1 -0.87148566         -0.8752239706       -0.875151653       \n",
       "LV2 -0.13233526         -0.1323391474       -0.131399397       \n",
       "LV3 -0.01513276         -0.0113543157       -0.024335924       \n",
       "LV4 -0.24443976         -0.2327881755       -0.236407836       \n",
       "LV5 -0.02083118         -0.0228384570       -0.017638770       \n",
       "LV6 -0.01517550          0.0009615722       -0.008850218       \n",
       "    SRP000599.SRR013555 SRP000599.SRR013556 SRP000599.SRR013557\n",
       "LV1 -0.82379528         -8.266395e-01       -0.82067562        \n",
       "LV2 -0.16725299         -1.608620e-01       -0.16321947        \n",
       "LV3 -0.03823130         -3.274484e-02       -0.04298509        \n",
       "LV4 -0.21912488         -2.010877e-01       -0.22754310        \n",
       "LV5 -0.04534587         -4.036835e-02       -0.03258473        \n",
       "LV6 -0.01707047         -7.413137e-05       -0.02293963        \n",
       "    SRP000599.SRR013558 â‹¯ SRP035599.SRR1139372 SRP035599.SRR1139393\n",
       "LV1 -0.82243867         â‹¯ -0.52053021          -0.51392282         \n",
       "LV2 -0.16428624         â‹¯  0.20624147           0.08443934         \n",
       "LV3 -0.04094039         â‹¯ -0.17209869          -0.16298852         \n",
       "LV4 -0.21997184         â‹¯ -0.09571376          -0.14086539         \n",
       "LV5 -0.03866460         â‹¯ -0.14597781          -0.14888982         \n",
       "LV6 -0.01355327         â‹¯ -0.24519268          -0.24951182         \n",
       "    SRP035599.SRR1139388 SRP035599.SRR1139378 SRP035599.SRR1139399\n",
       "LV1 -0.5257386           -0.5168343           -0.5285092          \n",
       "LV2  0.1507070            0.1775285            0.1510824          \n",
       "LV3 -0.1731993           -0.1882488           -0.1720055          \n",
       "LV4 -0.2011386           -0.1156240           -0.2145210          \n",
       "LV5 -0.1407341           -0.1453838           -0.1405794          \n",
       "LV6 -0.2527177           -0.2627610           -0.2605949          \n",
       "    SRP035599.SRR1139386 SRP035599.SRR1139375 SRP035599.SRR1139382\n",
       "LV1 -0.5401691           -0.5359329           -0.5174164          \n",
       "LV2  0.1872756            0.1662790            0.1755454          \n",
       "LV3 -0.1600504           -0.1761666           -0.1949913          \n",
       "LV4 -0.1652276           -0.2586825           -0.1696389          \n",
       "LV5 -0.1426756           -0.1224374           -0.1401182          \n",
       "LV6 -0.2633666           -0.2431627           -0.2478865          \n",
       "    SRP035599.SRR1139356 SRP035599.SRR1139370\n",
       "LV1 -0.5040608           -0.5383226          \n",
       "LV2  0.2382462            0.1948014          \n",
       "LV3 -0.1505919           -0.1967838          \n",
       "LV4 -0.1651986           -0.3765665          \n",
       "LV5 -0.1452800           -0.1319524          \n",
       "LV6 -0.2298372           -0.2663158          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "recount2_baseRes$Z <- data.frame(recount2_baseRes$Z)\n",
    "rownames(recount2_baseRes$Z) <- recount2_genes\n",
    "head(recount2_baseRes$Z)\n",
    "\n",
    "recount2_baseRes$B <- data.frame(recount2_baseRes$B)\n",
    "colnames(recount2_baseRes$B) <- samples\n",
    "head(recount2_baseRes$B)\n",
    "\n",
    "recount2_fullRes$summary <- recount2_fullRes$summary %>%\n",
    "    dplyr::rename(LV = LV_index)  %>% \n",
    "    dplyr::mutate(LV = paste0('LV', LV))\n",
    "\n",
    "saveRDS(recount2_fullRes, file = file.path(output_data_dir, \"CLAMPfull.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "model_dir <- file.path(output_data_dir, \"CLAMPfull\")\n",
    "dir.create(model_dir, showWarnings = FALSE, recursive = TRUE)\n",
    "\n",
    "B <- recount2_fullRes$B\n",
    "write.csv(B, file.path(model_dir, \"B.csv\"))\n",
    "\n",
    "Z <- recount2_fullRes$Z\n",
    "write.csv(Z, file.path(model_dir, \"Z.csv\"))\n",
    "\n",
    "summary <- recount2_fullRes$summary\n",
    "write.csv(summary, file.path(model_dir, \"summary.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Recount2 CLAMP analysis completed at: 2025-11-25 10:01:19 \n",
      "Total elapsed time: 21.08286 hours \n"
     ]
    }
   ],
   "source": [
    "# Create a timestamp to track the end of the analysis\n",
    "end_time <- Sys.time()\n",
    "elapsed_time <- end_time - start_time\n",
    "cat(\"Recount2 CLAMP analysis completed at:\", format(end_time), \"\\n\")\n",
    "cat(\"Total elapsed time:\", format(elapsed_time), \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "947ca517",
   "metadata": {},
   "source": [
    "# PLIER"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb463b2b",
   "metadata": {},
   "source": [
    "Run PLIER with the same inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95a907e4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Removing 0 pathways with too few genes\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 107.0666\n",
      "[1] \"L2 is set to 107.066621443032\"\n",
      "[1] \"L1 is set to 53.5333107215158\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "errorY (SVD based:best possible) = 0.692\n",
      "\n"
     ]
    }
   ],
   "source": [
    "recount2_plier = PLIER::PLIER(\n",
    "    recount2_fbm_filt[], \n",
    "    as.matrix(recount2_matched), \n",
    "    svdres = recount2_svdRes, \n",
    "    Chat = as.matrix(recount2_chatObj), \n",
    "    doCrossval = TRUE, \n",
    "    k = CLAMP_K_recount2\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 Ã— 37032 of type dbl</caption>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>LV 1</th><td> 3.038562e-11</td><td>-1.164697e-13</td><td> 1.499369e-11</td><td> 2.420048e-11</td><td> 1.076099e-11</td><td> 2.400729e-11</td><td> 2.450469e-11</td><td> 2.529211e-11</td><td> 2.492415e-11</td><td> 2.411303e-11</td><td>â‹¯</td><td>-2.252487e-14</td><td>-9.868426e-13</td><td> 7.459374e-13</td><td>-1.238694e-13</td><td>-5.471264e-13</td><td> 3.208637e-13</td><td>-1.203242e-12</td><td>-9.133111e-13</td><td>-7.087071e-13</td><td>-7.032934e-13</td></tr>\n",
       "\t<tr><th scope=row>2,Cells - Cultured Fibroblasts Female 70-79 Up</th><td>-3.281172e-02</td><td>-3.819620e-02</td><td>-4.021169e-02</td><td>-4.102633e-02</td><td>-3.925897e-02</td><td>-3.957653e-02</td><td>-9.111154e-03</td><td>-7.944571e-03</td><td>-1.111526e-02</td><td>-5.022869e-03</td><td>â‹¯</td><td> 3.445032e-02</td><td> 9.405888e-02</td><td> 5.244139e-02</td><td> 9.127825e-02</td><td> 8.897180e-02</td><td> 6.534399e-02</td><td> 7.464888e-02</td><td> 1.037498e-01</td><td> 4.380141e-02</td><td> 9.150614e-02</td></tr>\n",
       "\t<tr><th scope=row>3,mRNA Splicing, via Spliceosome (GO:0000398)</th><td>-9.481710e-02</td><td>-7.740678e-02</td><td>-9.260198e-02</td><td>-8.277387e-02</td><td>-7.081893e-02</td><td>-8.502113e-02</td><td>-6.194781e-02</td><td>-6.563719e-02</td><td>-6.688297e-02</td><td>-7.041583e-02</td><td>â‹¯</td><td>-5.538873e-02</td><td>-6.185108e-02</td><td>-3.681471e-02</td><td>-3.960329e-02</td><td>-6.117308e-02</td><td>-2.970421e-02</td><td>-8.491269e-02</td><td>-3.287660e-02</td><td>-2.841848e-02</td><td>-1.169766e-01</td></tr>\n",
       "\t<tr><th scope=row>4,Translation (GO:0006412)</th><td>-1.167416e-01</td><td>-1.169561e-01</td><td>-1.241410e-01</td><td>-1.144214e-01</td><td>-1.081250e-01</td><td>-1.183666e-01</td><td>-1.348015e-01</td><td>-1.444428e-01</td><td>-1.385527e-01</td><td>-1.426251e-01</td><td>â‹¯</td><td> 5.650564e-02</td><td> 2.639246e-02</td><td> 1.109185e-01</td><td> 6.937041e-02</td><td> 8.791064e-02</td><td> 1.055854e-01</td><td> 1.084018e-01</td><td> 7.893540e-02</td><td> 7.554681e-02</td><td> 1.749875e-01</td></tr>\n",
       "\t<tr><th scope=row>5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)</th><td>-1.475776e-02</td><td>-2.136897e-02</td><td>-2.183746e-02</td><td>-1.868801e-02</td><td>-1.880394e-02</td><td>-1.911732e-02</td><td>-2.733048e-02</td><td>-3.572729e-02</td><td>-3.150585e-02</td><td>-2.717683e-02</td><td>â‹¯</td><td> 4.954903e-03</td><td> 2.586807e-02</td><td> 2.185875e-02</td><td> 4.275850e-02</td><td> 4.369622e-02</td><td> 5.298591e-02</td><td>-1.144091e-02</td><td> 5.217841e-02</td><td>-1.841314e-02</td><td> 2.890525e-02</td></tr>\n",
       "\t<tr><th scope=row>LV 6</th><td> 1.137411e-02</td><td> 1.100171e-02</td><td> 2.021871e-02</td><td> 2.901634e-02</td><td> 1.232140e-02</td><td> 1.205028e-02</td><td> 9.152861e-03</td><td> 1.060361e-02</td><td>-6.413729e-04</td><td> 6.352040e-03</td><td>â‹¯</td><td>-7.302503e-02</td><td>-5.922457e-02</td><td>-7.198972e-02</td><td>-5.402172e-02</td><td>-6.231897e-02</td><td>-6.095302e-02</td><td>-5.806830e-02</td><td>-6.060223e-02</td><td>-7.126057e-02</td><td>-5.891329e-02</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 Ã— 37032 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "\tLV 1 &  3.038562e-11 & -1.164697e-13 &  1.499369e-11 &  2.420048e-11 &  1.076099e-11 &  2.400729e-11 &  2.450469e-11 &  2.529211e-11 &  2.492415e-11 &  2.411303e-11 & â‹¯ & -2.252487e-14 & -9.868426e-13 &  7.459374e-13 & -1.238694e-13 & -5.471264e-13 &  3.208637e-13 & -1.203242e-12 & -9.133111e-13 & -7.087071e-13 & -7.032934e-13\\\\\n",
       "\t2,Cells - Cultured Fibroblasts Female 70-79 Up & -3.281172e-02 & -3.819620e-02 & -4.021169e-02 & -4.102633e-02 & -3.925897e-02 & -3.957653e-02 & -9.111154e-03 & -7.944571e-03 & -1.111526e-02 & -5.022869e-03 & â‹¯ &  3.445032e-02 &  9.405888e-02 &  5.244139e-02 &  9.127825e-02 &  8.897180e-02 &  6.534399e-02 &  7.464888e-02 &  1.037498e-01 &  4.380141e-02 &  9.150614e-02\\\\\n",
       "\t3,mRNA Splicing, via Spliceosome (GO:0000398) & -9.481710e-02 & -7.740678e-02 & -9.260198e-02 & -8.277387e-02 & -7.081893e-02 & -8.502113e-02 & -6.194781e-02 & -6.563719e-02 & -6.688297e-02 & -7.041583e-02 & â‹¯ & -5.538873e-02 & -6.185108e-02 & -3.681471e-02 & -3.960329e-02 & -6.117308e-02 & -2.970421e-02 & -8.491269e-02 & -3.287660e-02 & -2.841848e-02 & -1.169766e-01\\\\\n",
       "\t4,Translation (GO:0006412) & -1.167416e-01 & -1.169561e-01 & -1.241410e-01 & -1.144214e-01 & -1.081250e-01 & -1.183666e-01 & -1.348015e-01 & -1.444428e-01 & -1.385527e-01 & -1.426251e-01 & â‹¯ &  5.650564e-02 &  2.639246e-02 &  1.109185e-01 &  6.937041e-02 &  8.791064e-02 &  1.055854e-01 &  1.084018e-01 &  7.893540e-02 &  7.554681e-02 &  1.749875e-01\\\\\n",
       "\t5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) & -1.475776e-02 & -2.136897e-02 & -2.183746e-02 & -1.868801e-02 & -1.880394e-02 & -1.911732e-02 & -2.733048e-02 & -3.572729e-02 & -3.150585e-02 & -2.717683e-02 & â‹¯ &  4.954903e-03 &  2.586807e-02 &  2.185875e-02 &  4.275850e-02 &  4.369622e-02 &  5.298591e-02 & -1.144091e-02 &  5.217841e-02 & -1.841314e-02 &  2.890525e-02\\\\\n",
       "\tLV 6 &  1.137411e-02 &  1.100171e-02 &  2.021871e-02 &  2.901634e-02 &  1.232140e-02 &  1.205028e-02 &  9.152861e-03 &  1.060361e-02 & -6.413729e-04 &  6.352040e-03 & â‹¯ & -7.302503e-02 & -5.922457e-02 & -7.198972e-02 & -5.402172e-02 & -6.231897e-02 & -6.095302e-02 & -5.806830e-02 & -6.060223e-02 & -7.126057e-02 & -5.891329e-02\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 Ã— 37032 of type dbl\n",
       "\n",
       "| LV 1 |  3.038562e-11 | -1.164697e-13 |  1.499369e-11 |  2.420048e-11 |  1.076099e-11 |  2.400729e-11 |  2.450469e-11 |  2.529211e-11 |  2.492415e-11 |  2.411303e-11 | â‹¯ | -2.252487e-14 | -9.868426e-13 |  7.459374e-13 | -1.238694e-13 | -5.471264e-13 |  3.208637e-13 | -1.203242e-12 | -9.133111e-13 | -7.087071e-13 | -7.032934e-13 |\n",
       "| 2,Cells - Cultured Fibroblasts Female 70-79 Up | -3.281172e-02 | -3.819620e-02 | -4.021169e-02 | -4.102633e-02 | -3.925897e-02 | -3.957653e-02 | -9.111154e-03 | -7.944571e-03 | -1.111526e-02 | -5.022869e-03 | â‹¯ |  3.445032e-02 |  9.405888e-02 |  5.244139e-02 |  9.127825e-02 |  8.897180e-02 |  6.534399e-02 |  7.464888e-02 |  1.037498e-01 |  4.380141e-02 |  9.150614e-02 |\n",
       "| 3,mRNA Splicing, via Spliceosome (GO:0000398) | -9.481710e-02 | -7.740678e-02 | -9.260198e-02 | -8.277387e-02 | -7.081893e-02 | -8.502113e-02 | -6.194781e-02 | -6.563719e-02 | -6.688297e-02 | -7.041583e-02 | â‹¯ | -5.538873e-02 | -6.185108e-02 | -3.681471e-02 | -3.960329e-02 | -6.117308e-02 | -2.970421e-02 | -8.491269e-02 | -3.287660e-02 | -2.841848e-02 | -1.169766e-01 |\n",
       "| 4,Translation (GO:0006412) | -1.167416e-01 | -1.169561e-01 | -1.241410e-01 | -1.144214e-01 | -1.081250e-01 | -1.183666e-01 | -1.348015e-01 | -1.444428e-01 | -1.385527e-01 | -1.426251e-01 | â‹¯ |  5.650564e-02 |  2.639246e-02 |  1.109185e-01 |  6.937041e-02 |  8.791064e-02 |  1.055854e-01 |  1.084018e-01 |  7.893540e-02 |  7.554681e-02 |  1.749875e-01 |\n",
       "| 5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) | -1.475776e-02 | -2.136897e-02 | -2.183746e-02 | -1.868801e-02 | -1.880394e-02 | -1.911732e-02 | -2.733048e-02 | -3.572729e-02 | -3.150585e-02 | -2.717683e-02 | â‹¯ |  4.954903e-03 |  2.586807e-02 |  2.185875e-02 |  4.275850e-02 |  4.369622e-02 |  5.298591e-02 | -1.144091e-02 |  5.217841e-02 | -1.841314e-02 |  2.890525e-02 |\n",
       "| LV 6 |  1.137411e-02 |  1.100171e-02 |  2.021871e-02 |  2.901634e-02 |  1.232140e-02 |  1.205028e-02 |  9.152861e-03 |  1.060361e-02 | -6.413729e-04 |  6.352040e-03 | â‹¯ | -7.302503e-02 | -5.922457e-02 | -7.198972e-02 | -5.402172e-02 | -6.231897e-02 | -6.095302e-02 | -5.806830e-02 | -6.060223e-02 | -7.126057e-02 | -5.891329e-02 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                                       [,1]         \n",
       "LV 1                                                                    3.038562e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -3.281172e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -9.481710e-02\n",
       "4,Translation (GO:0006412)                                             -1.167416e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.475776e-02\n",
       "LV 6                                                                    1.137411e-02\n",
       "                                                                       [,2]         \n",
       "LV 1                                                                   -1.164697e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -3.819620e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -7.740678e-02\n",
       "4,Translation (GO:0006412)                                             -1.169561e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -2.136897e-02\n",
       "LV 6                                                                    1.100171e-02\n",
       "                                                                       [,3]         \n",
       "LV 1                                                                    1.499369e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -4.021169e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -9.260198e-02\n",
       "4,Translation (GO:0006412)                                             -1.241410e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -2.183746e-02\n",
       "LV 6                                                                    2.021871e-02\n",
       "                                                                       [,4]         \n",
       "LV 1                                                                    2.420048e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -4.102633e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -8.277387e-02\n",
       "4,Translation (GO:0006412)                                             -1.144214e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.868801e-02\n",
       "LV 6                                                                    2.901634e-02\n",
       "                                                                       [,5]         \n",
       "LV 1                                                                    1.076099e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -3.925897e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -7.081893e-02\n",
       "4,Translation (GO:0006412)                                             -1.081250e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.880394e-02\n",
       "LV 6                                                                    1.232140e-02\n",
       "                                                                       [,6]         \n",
       "LV 1                                                                    2.400729e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -3.957653e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -8.502113e-02\n",
       "4,Translation (GO:0006412)                                             -1.183666e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.911732e-02\n",
       "LV 6                                                                    1.205028e-02\n",
       "                                                                       [,7]         \n",
       "LV 1                                                                    2.450469e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -9.111154e-03\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -6.194781e-02\n",
       "4,Translation (GO:0006412)                                             -1.348015e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -2.733048e-02\n",
       "LV 6                                                                    9.152861e-03\n",
       "                                                                       [,8]         \n",
       "LV 1                                                                    2.529211e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -7.944571e-03\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -6.563719e-02\n",
       "4,Translation (GO:0006412)                                             -1.444428e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -3.572729e-02\n",
       "LV 6                                                                    1.060361e-02\n",
       "                                                                       [,9]         \n",
       "LV 1                                                                    2.492415e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -1.111526e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -6.688297e-02\n",
       "4,Translation (GO:0006412)                                             -1.385527e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -3.150585e-02\n",
       "LV 6                                                                   -6.413729e-04\n",
       "                                                                       [,10]        \n",
       "LV 1                                                                    2.411303e-11\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         -5.022869e-03\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -7.041583e-02\n",
       "4,Translation (GO:0006412)                                             -1.426251e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -2.717683e-02\n",
       "LV 6                                                                    6.352040e-03\n",
       "                                                                       [,11]\n",
       "LV 1                                                                   â‹¯    \n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                         â‹¯    \n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          â‹¯    \n",
       "4,Translation (GO:0006412)                                             â‹¯    \n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) â‹¯    \n",
       "LV 6                                                                   â‹¯    \n",
       "                                                                       [,12]        \n",
       "LV 1                                                                   -2.252487e-14\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          3.445032e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -5.538873e-02\n",
       "4,Translation (GO:0006412)                                              5.650564e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  4.954903e-03\n",
       "LV 6                                                                   -7.302503e-02\n",
       "                                                                       [,13]        \n",
       "LV 1                                                                   -9.868426e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          9.405888e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -6.185108e-02\n",
       "4,Translation (GO:0006412)                                              2.639246e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  2.586807e-02\n",
       "LV 6                                                                   -5.922457e-02\n",
       "                                                                       [,14]        \n",
       "LV 1                                                                    7.459374e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          5.244139e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -3.681471e-02\n",
       "4,Translation (GO:0006412)                                              1.109185e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  2.185875e-02\n",
       "LV 6                                                                   -7.198972e-02\n",
       "                                                                       [,15]        \n",
       "LV 1                                                                   -1.238694e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          9.127825e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -3.960329e-02\n",
       "4,Translation (GO:0006412)                                              6.937041e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  4.275850e-02\n",
       "LV 6                                                                   -5.402172e-02\n",
       "                                                                       [,16]        \n",
       "LV 1                                                                   -5.471264e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          8.897180e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -6.117308e-02\n",
       "4,Translation (GO:0006412)                                              8.791064e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  4.369622e-02\n",
       "LV 6                                                                   -6.231897e-02\n",
       "                                                                       [,17]        \n",
       "LV 1                                                                    3.208637e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          6.534399e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -2.970421e-02\n",
       "4,Translation (GO:0006412)                                              1.055854e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  5.298591e-02\n",
       "LV 6                                                                   -6.095302e-02\n",
       "                                                                       [,18]        \n",
       "LV 1                                                                   -1.203242e-12\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          7.464888e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -8.491269e-02\n",
       "4,Translation (GO:0006412)                                              1.084018e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.144091e-02\n",
       "LV 6                                                                   -5.806830e-02\n",
       "                                                                       [,19]        \n",
       "LV 1                                                                   -9.133111e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          1.037498e-01\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -3.287660e-02\n",
       "4,Translation (GO:0006412)                                              7.893540e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  5.217841e-02\n",
       "LV 6                                                                   -6.060223e-02\n",
       "                                                                       [,20]        \n",
       "LV 1                                                                   -7.087071e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          4.380141e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -2.841848e-02\n",
       "4,Translation (GO:0006412)                                              7.554681e-02\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724) -1.841314e-02\n",
       "LV 6                                                                   -7.126057e-02\n",
       "                                                                       [,21]        \n",
       "LV 1                                                                   -7.032934e-13\n",
       "2,Cells - Cultured Fibroblasts Female 70-79 Up                          9.150614e-02\n",
       "3,mRNA Splicing, via Spliceosome (GO:0000398)                          -1.169766e-01\n",
       "4,Translation (GO:0006412)                                              1.749875e-01\n",
       "5,Double-Strand Break Repair via Homologous Recombination (GO:0000724)  2.890525e-02\n",
       "LV 6                                                                   -5.891329e-02"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(gtex_plier$Z) <- paste0('LV', seq_len(ncol(gtex_plier$Z)))\n",
    "\n",
    "gtex_plier$summary <- gtex_plier$summary %>%\n",
    "    dplyr::rename(LV = `LV index`)  %>% \n",
    "    dplyr::mutate(LV = paste0('LV', LV))\n",
    "\n",
    "colnames(gtex_plier$B) <- samples\n",
    "\n",
    "saveRDS(gtex_plier, file = file.path(output_data_dir, \"PLIER.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "model_dir <- file.path(output_data_dir, \"PLIER\")\n",
    "dir.create(model_dir, showWarnings = FALSE, recursive = TRUE)\n",
    "\n",
    "B <- gtex_plier$B\n",
    "write.csv(B, file.path(model_dir, \"B.csv\"))\n",
    "\n",
    "Z <- gtex_plier$Z\n",
    "write.csv(Z, file.path(model_dir, \"Z.csv\"))\n",
    "\n",
    "summary <- gtex_plier$summary\n",
    "write.csv(summary, file.path(model_dir, \"summary.csv\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
