{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ca81bc7f",
   "metadata": {},
   "source": [
    "# GTEx model building with flashier"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4176d4ba-4225-43df-88a3-d47b01944c55",
   "metadata": {},
   "source": [
    "Empirical Bayes non-negative matrix factorization for single-cell RNA-seq data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a0f965d",
   "metadata": {},
   "source": [
    "## Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a81c183-1565-4b60-b5ef-3c985fbee80e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading GitHub repo willwerscheid/flashier@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ebnm         (NA       -> 825edf1e3...) [GitHub]\n",
      "xfun         (0.52     -> 0.53        ) [CRAN]\n",
      "evaluate     (1.0.4    -> 1.0.5       ) [CRAN]\n",
      "rmarkdown    (2.29     -> 2.30        ) [CRAN]\n",
      "stringr      (1.5.1    -> 1.5.2       ) [CRAN]\n",
      "later        (1.4.2    -> 1.4.4       ) [CRAN]\n",
      "openssl      (2.3.3    -> 2.3.4       ) [CRAN]\n",
      "curl         (6.2.2    -> 7.0.0       ) [CRAN]\n",
      "pillar       (1.11.0   -> 1.11.1      ) [CRAN]\n",
      "promises     (1.3.2    -> 1.3.3       ) [CRAN]\n",
      "crosstalk    (1.2.1    -> 1.2.2       ) [CRAN]\n",
      "magrittr     (2.0.3    -> 2.0.4       ) [CRAN]\n",
      "ggplot2      (3.5.2    -> 4.0.0       ) [CRAN]\n",
      "RcppArmad... (14.4.2-1 -> 15.0.2-2    ) [CRAN]\n",
      "invgamma     (NA       -> 1.2         ) [CRAN]\n",
      "etrunct      (NA       -> 0.1         ) [CRAN]\n",
      "mixsqp       (NA       -> 0.3-54      ) [CRAN]\n",
      "truncnorm    (NA       -> 1.0-9       ) [CRAN]\n",
      "dqrng        (0.3.2    -> 0.4.1       ) [CRAN]\n",
      "cowplot      (1.1.3    -> 1.2.0       ) [CRAN]\n",
      "RcppParallel (NA       -> 5.1.11-1    ) [CRAN]\n",
      "ashr         (NA       -> 2.2-63      ) [CRAN]\n",
      "quadprog     (NA       -> 1.5-8       ) [CRAN]\n",
      "deconvolveR  (NA       -> 1.2-1       ) [CRAN]\n",
      "horseshoe    (NA       -> 0.2.0       ) [CRAN]\n",
      "trust        (NA       -> 0.1-8       ) [CRAN]\n",
      "fastTopics   (NA       -> 0.6-192     ) [CRAN]\n",
      "softImpute   (NA       -> 1.4-3       ) [CRAN]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Skipping 1 packages not available: Polychrome\n",
      "\n",
      "Downloading GitHub repo stephenslab/ebnm@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pillar       (1.11.0   -> 1.11.1  ) [CRAN]\n",
      "magrittr     (2.0.3    -> 2.0.4   ) [CRAN]\n",
      "RcppArmad... (14.4.2-1 -> 15.0.2-2) [CRAN]\n",
      "invgamma     (NA       -> 1.2     ) [CRAN]\n",
      "etrunct      (NA       -> 0.1     ) [CRAN]\n",
      "mixsqp       (NA       -> 0.3-54  ) [CRAN]\n",
      "truncnorm    (NA       -> 1.0-9   ) [CRAN]\n",
      "ggplot2      (3.5.2    -> 4.0.0   ) [CRAN]\n",
      "deconvolveR  (NA       -> 1.2-1   ) [CRAN]\n",
      "trust        (NA       -> 0.1-8   ) [CRAN]\n",
      "ashr         (NA       -> 2.2-63  ) [CRAN]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing 11 packages: pillar, magrittr, RcppArmadillo, invgamma, etrunct, mixsqp, truncnorm, ggplot2, deconvolveR, trust, ashr\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!requireNamespace(\"flashier\", quietly = TRUE)) {\n",
    "  remotes::install_github(\"willwerscheid/flashier\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2af9c3b3-a074-4af6-8c25-346823afe1bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(flashier)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9183afc-1007-4680-8144-ce00c0b72923",
   "metadata": {},
   "outputs": [],
   "source": [
    "data(gtex)\n",
    "flash_n <- flash(gtex, ebnm_fn = ebnm_normal, backfit = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a7d42a51",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in library(flashier): there is no package called ‘flashier’\n",
     "output_type": "error",
     "traceback": [
      "Error in library(flashier): there is no package called ‘flashier’\nTraceback:\n",
      "1. stop(packageNotFoundError(package, lib.loc, sys.call()))"
     ]
    }
   ],
   "source": [
    "library(flashier)\n",
    "library(Matrix)\n",
    "library(fastTopics)\n",
    "library(ggplot2)\n",
    "set.seed(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb7155e7",
   "metadata": {},
   "source": [
    "## Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c639a250",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in here(\"output/gtex/df_gtex_fbm_filt.rds\"): could not find function \"here\"\n",
     "output_type": "error",
     "traceback": [
      "Error in here(\"output/gtex/df_gtex_fbm_filt.rds\"): could not find function \"here\"\nTraceback:\n",
      "1. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = NULL)\n . }, \"could not find function \\\"here\\\"\", base::quote(here(\"output/gtex/df_gtex_fbm_filt.rds\")))"
     ]
    }
   ],
   "source": [
    "gtex_data <- readRDS(here('output/gtex/df_gtex_fbm_filt.rds'))\n",
    "head(gtex_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "487bf0c9-f978-4b6e-a8a9-41071297d71e",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts <- pbmc_facs$counts\n",
    "colnames(counts) <- make.names(pbmc_facs$genes$symbol, unique = TRUE)\n",
    "dim(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bceedad2-627c-4eb5-a229-4a2ca783b8f4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:plier2-analyses]",
   "language": "R",
   "name": "conda-env-plier2-analyses-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
